---
title: "Colletotrichum_H109_MAKER_annotation"
author: "Emily Giroux"
date: "6/11/2019"
output: pdf_document
fontsize: 11pt
geometry: margin=1in
urlcolor: blue
header-includes: \usepackage{xcolor}
---
```{r, global_options, eval=TRUE, echo=FALSE, cache=TRUE}
#Set the global options for knitr
library(knitr)
opts_chunk$set(tidy.opts=list(width.cutoff=80), tidy = TRUE, fig.align='center',
               cache=FALSE, collapse=TRUE, echo=FALSE, eval=FALSE, include=FALSE,
               message=FALSE, quietly=TRUE, results='hide', warn.conflicts=FALSE, 
               warning=FALSE)
```

```{r, installation1, eval=TRUE, echo=FALSE, include=FALSE, message=FALSE, warning=FALSE, cache=TRUE}
#Installing required packages
r = getOption("repos")
r["CRAN"] = "http://cran.us.r-project.org"
options(repos = r)

if(!require(devtools)) install.packages("devtools")
devtools::install_github("kassambara/fastqcr")

if (!requireNamespace("BiocManager"))
    install.packages("BiocManager")
BiocManager::install()

library("BiocManager")
.cran_packages <- c("ggplot2", "gridExtra", "rprojroot", "data.table", 
                    "knitr", "kableExtra", "cowplot", "filesstrings",
                    "tidyr", "reshape2", "kableExtra")
.bioc_packages <- c("BiocStyle", "Biostrings")
.inst <- .cran_packages %in% installed.packages()
if(any(!.inst)) {
   install.packages(.cran_packages[!.inst])
}
.inst <- .bioc_packages %in% installed.packages()
if(any(!.inst)) {
  BiocManager::install(.bioc_packages[!.inst], ask = FALSE)
}
sapply(c(.cran_packages, .bioc_packages), require, character.only = TRUE)
```

```{r sourcing_my_functions, echo=FALSE, eval=TRUE, include=FALSE, cache=TRUE}
#Source our custom R scripts:    
#For this we will use the rprojroot package to set the directory structures. This will help us when finding our files to source functions. We specify ours is an RStudio project. The root object contains a function that will help us locate our package R files regarless of our current working directory.
library(rprojroot)
root <- rprojroot::is_rstudio_project
scriptsPath <- root$make_fix_file(".")("R")
scripts  <- dir(root$find_file("R", path = root$find_file()))
scriptsl <- paste(scriptsPath, scripts, sep = "//")
lapply(scriptsl, source)
# Record the path to the environment images directory:
sharedPath <- "/isilon/cfia-ottawa-fallowfield/users/girouxeml/PIRL_working_directory/"
analysis   <- "colletotrichum_H109_annotation/"
sharedPathAn <- paste(sharedPath, analysis, sep = "")

imageDirPath <- "/isilon/cfia-ottawa-fallowfield/users/girouxeml/GitHub_Repos/r_environments/colletotrichum_H109_annotation/"
baseImage    <- "colletotrichum_H109_annotation_June2019.RData"

save.image(paste(imageDirPath, baseImage, sep = ""))
load(paste(imageDirPath, baseImage, sep = ""))
```

User:
The following paths are to directories where the references, tools and general 
requirements are located, this depends on the directories actually having been 
put there:
```{r user_tools_and_references_dir, cache=TRUE}
referencesPath   <- paste(sharedPath, "References/", sep = "")
```

```{r analysis_and_sequence_directory, cache=TRUE}
analysis   <- "Lachnellula_species_GenomeAn_IonTorrent_2017/"
seqDataDir <- "IonTorrent_data_2017"
# Set the path to the analysis directory:
sharedPathAn <- paste(sharedPath, analysis, sep="")

# Set the path for where I've been putting the NCBI sumission files:
lsueNCBIfilesPath <- paste(sharedPathAn, "Lachnellula_suecica/Lsue_NCBI_files/Lsue_NCBI/", sep = "")
```

Set paths to Maker and GAG scripts:
```{r}
# Path of where I'm saving scripts:
scriptsPath <- "/home/CFIA-ACIA/girouxeml/prog/scripts_pl/"

# Maker Scripts:
makerMapPath <- paste(scriptsPath, "maker_map_ids.pl", sep = "")
makerMapGffPath <- paste(scriptsPath, "map_gff_ids.pl", sep = "")
makerFuncGffPath <- paste(scriptsPath, "maker_functional_gff.pl", sep = "")
makerIPRupdatePath <- paste(scriptsPath, "ipr_update_gff.pl", sep = "")
makerDatMapPath <- paste(scriptsPath, "map_data_ids.pl", sep = "")
makerMapFastaPath <- paste(scriptsPath, "map_fasta_ids.pl", sep = "")

# GAG scripts: 
gagPath <- "/home/CFIA-ACIA/girouxeml/prog/gag/genomeannotation-GAG-40ea515/gag.py"

# Annie Scripts:
anniePath <- paste("/home/CFIA-ACIA/girouxeml/prog/annie/genomeannotation-annie/annie.py")
```

Create output directories for Maker and record the path:
```{r}
dir.create(paste(lsueNCBIfilesPath, "Maker_Output", sep = ""), showWarnings = TRUE, recursive = FALSE)
makerOutput <- paste(lsueNCBIfilesPath, "Maker_Output/", sep = "")
```


Record the path of relevant files such as assembly, gff, and annotation files:
```{r}
# GFF exported from Blast2GO, following Blast2GO geneFind for eukaryotes:
gffPath <- paste(lsueNCBIfilesPath, "B2G_project_files/Blast2GO_gff_without_contig.gff", sep = "")

# Assembly fasta file from Newbler:
fnaAssPath <- paste(lsueNCBIfilesPath, 
                    "Assembly_Files/FromB2GSupport_Lsue_assembly_EG_2017_mod_removedlLessThan200_contig00022trim.fna", sep = "")
```


Copy the gff file to the new Maker_Output directory and record path to the copied gff:
```{r}
file.copy(gffPath, makerOutput)
gffPathPr <- paste(makerOutput, "Blast2GO_gff_without_contig.gff", sep = "")
```
** Over here:
Testing: I opened the blast gff in gedit, and changed all instances of seqName to Name.

Seems like there are issues with the format og the Blast2GO gff, namely,
the attributes use "seqName" and then re-order the other ones. I am using Maker
to add annotations, so reformat to look like Maker GFF before using Maker scripts
```{r}
gff <- read.table(file = gffPathPr, 
                  sep = "\t", header = F, quote = "", comment.char="#", fill = T)

gff_c9 <- subset(gff, select = c(V1,V2,V3,V9))
gff_c9$v1 <- as.character(lapply(strsplit(as.character(gff_c9$V9), split = ";"), "[",1))
gff_c9$v2 <- as.character(lapply(strsplit(as.character(gff_c9$V9), split = "="), "[",2))
gff_c9$v2 <- gsub(";.*", "", gff_c9$v2)
gff_c9$v3 <- as.character(lapply(strsplit(as.character(gff_c9$V9), split = ";"), "[",2))
gff_c9$v4 <- as.character(lapply(strsplit(as.character(gff_c9$V9), split = ";"), "[",3))

setDT(gff_c9)
gff_c9[ V3 == "gene", v1 := paste("ID=",v2, sep = "")]
gff_c9[ V3 == "gene", v3 := paste("Name=",v2, sep = "")]
gff_c9[ V3 == "gene", v5 := paste(v1, v3, sep = ";")]

gff_c9[ V3 == "mRNA", v3 := paste("Name=",v2, sep = "")]
gff_c9[ V3 == "mRNA", v5 := paste(v4, v1, v3, sep = ";")]

gff_c9[ V3 == "intron", v3 := gsub("seqName", "Name", gff_c9$v3)]
gff_c9[ V3 == "intron", v5 := paste(v1, v3, sep = ";")]

gff_c9[ V3 == "stop_codon", v3 := gsub("seqName", "Name", gff_c9$v3)]
gff_c9[ V3 == "stop_codon", v5 := paste(v1, v3, sep = ";")]

gff_c9[ V3 == "start_codon", v3 := gsub("seqName", "Name", gff_c9$v3)]
gff_c9[ V3 == "start_codon", v5 := paste(v1, v3, sep = ";")]

gff_c9[ V3 == "CDS", v3 := gsub("seqName", "Name", gff_c9$v3)]
gff_c9[ V3 == "CDS", v5 := paste(v1, v3, v4, sep = ";")]

gff$V9 <- gff_c9$v5

write.table(gff, file = paste(makerOutput, "mod_gff.gff", sep = ""), 
      append = FALSE,
      sep = "\t",
      col.names=FALSE, row.names = FALSE, quote = FALSE)
```

Stack overflow - add the contig lines to the mod2_gff.gff file, using the terminal:
$ awk 'p!=$1 {print "# Fasta definition line: >" $1;
print "##sequence-region";
p=$1}1' mod_gff.gff > mod2_gff.gff

After this is done, open the gff in gedit, and add the fasta definition lines 1-7 that are in 
the original blast2go gff.
```{r}
# record the path tot he modified GFF
modGFF <- paste(makerOutput, "mod2_gff.gff", sep = "")
```

Rename gene names from gff and other output files for annotations:
Map ID file:
```{r}
prefix = "LSUE_"
justify = 6
abrvGene = "G"
abrvTrans = "T"

mapFilePath <- paste(makerOutput, "id_Lsue2.map", sep = "")

cmd <- paste("perl ", makerMapPath,
             " --prefix ", prefix,
             " --justify ", justify,
             " --abrv_gene ", abrvGene,
             " --abrv_tran ", abrvTrans,
             " ", modGFF, # or try gffPathPr
             " > ", mapFilePath, sep = "")

system(cmd)
```

In-place replacement of names in GFF file with Maker scripts using the mapping file created
in previous chunk:
```{r}
cmd <- paste("perl ", makerMapGffPath, 
            " ", mapFilePath, 
            " ", modGFF, # or try gffPathPr
            sep = "")

system(cmd)
```

```{r}
uniprot_uniparc_union <- paste(makerOutput, "sp_Union_complete_db.fa", sep = "")
blast_union_uniprot_uniparc <- paste(makerOutput, "sp_Union_completeBlast_db.txt", sep = "")

gffRenamedFuncPath_union <- paste(makerOutput, "func_Union_Renamed_Blast2GO_gff_without_contig.gff", sep = "")

cmd <- paste("perl ", makerFuncGffPath,
             " ", uniprot_uniparc_union,
             " ", blast_union_uniprot_uniparc, 
             " ", modGFF, # or try gffPathPr
             " > ", gffRenamedFuncPath_union, sep = "")

system(cmd)
```


Path to Genome Annotation GitHub (GAG) tool (python):
```{r}
gagPath <- "/home/CFIA-ACIA/girouxeml/prog/gag/genomeannotation-GAG-40ea515/gag.py"

testGAG <- paste("python ", gagPath, " --help", sep = "")
system(testGAG)
```

Import the gff, fna, and annotation files to inspect their formatting:
```{r}
library(data.table)
library("dplyr")

fnaPath <- paste(sharedPathAn, 
                 "Lachnellula_suecica/Lsue_NCBI_files/Lsue_NCBI/FromB2GSupport_Lsue_assembly_EG_2017_mod_removedlLessThan200_contig00022trim.fna", 
                 sep = "") 
gffPath <- paste(sharedPathAn, "Lachnellula_suecica/Lsue_NCBI_files/Lsue_NCBI/Blast2GO_gff_without_contig.gff", sep = "")
annotPath <- paste(sharedPathAn, "Lachnellula_suecica/Lsue_NCBI_files/Lsue_NCBI/Lsue_blast2go_annot_05Oct2017_1059.annot", 
                   sep = "")
annotDesPath <- paste(sharedPathAn, "Lachnellula_suecica/Lsue_NCBI_files/Lsue_NCBI/Lsue_blast2go_annotation_descriptions_05Oct2017_1128.txt", 
                      sep = "")
gffExportPath <- paste(sharedPathAn, "Lachnellula_suecica/Lsue_NCBI_files/Lsue_NCBI/blast2go_gff_export_05Oct2017_1132.gff", 
                       sep = "")
gafPath <- paste(sharedPathAn, "Lachnellula_suecica/Lsue_NCBI_files/Lsue_NCBI/Lsue_blast2go_gaf_05Oct2017_1227.txt", sep = "")


gff <- fread(gffPath, sep = "\t", header = FALSE)
annot <- fread(annotPath, sep = "\t", header = FALSE, fill = TRUE)
annotDes <- fread(annotDesPath, sep = "auto", header = TRUE, fill=TRUE)
gffExport <- fread(gffExportPath, sep = "\t", header = FALSE)
gaf <- fread(gafPath, sep = "\t", header = FALSE)

library(Biostrings)
fna <- readDNAStringSet(fnaPath)


makerMapIDsPath <- "/home/CFIA-ACIA/girouxeml/prog/scripts_pl/maker_map_ids.pl"
makerMapFastaIdsPath <- "/home/CFIA-ACIA/girouxeml/prog/scripts_pl/map_fasta_ids.pl"
makerMapGffIdsPath <- "/home/CFIA-ACIA/girouxeml/prog/scripts_pl/map_gff_ids.pl"    
makerMapDataIdsPath <- "/home/CFIA-ACIA/girouxeml/prog/scripts_pl/map_data_ids.pl"

testMakerMapIds <- paste("perl ", makerMapGffIdsPath, sep = "")
system(testMakerMapIds)

# Step 1: make an ID map file
prefixIDs <- "LSUE_"
justify <- 6

cmd <- paste("perl ", makerMapIDsPath, 
             " --prefix ", prefixIDs,
             #" --suffix ", 
             " --abrv_gene G ",
             " --abrv_tran T ",
             " --justify ", justify,
             " ", gffPath, " > ",
             paste(sharedPathAn, "Lachnellula_suecica/Lsue_NCBI_files/Lsue_NCBI/Maker_Map_Lsue.map", sep = ""),
             sep = "")

system(cmd)

# Step 2: Map gff ids
cmd <- paste("cp ", gffPath, " ", 
             paste(sharedPathAn, "Lachnellula_suecica/Lsue_NCBI_files/Lsue_NCBI/contig00047_Lsue_Renamed.gff", sep = ""),
             sep = "")
system(cmd)

cmd <- paste("perl ", makerMapGffIdsPath, " ",
             paste(sharedPathAn, "Lachnellula_suecica/Lsue_NCBI_files/Lsue_NCBI/Maker_map_contig00047_Lsue.map", sep = ""),
             " ", 
             paste(sharedPathAn, "Lachnellula_suecica/Lsue_NCBI_files/Lsue_NCBI/contig00047_Lsue_Renamed.gff", sep = ""),
             sep = "")

system(cmd)

# Step 3: Map fasta IDs
b2gAAFasta <- "/home/CFIA-ACIA/girouxeml/PIRL_working_directory/Lachnellula_species_GenomeAn_IonTorrent_2017/Lachnellula_suecica/Lsue_NCBI_files/Lsue_NCBI/Lsue_blast2go_fasta_05Oct2017_1535.fasta"

cmd <- paste("cp ", b2gAAFasta, " ", 
             paste(sharedPathAn, "Lachnellula_suecica/Lsue_NCBI_files/Lsue_NCBI/Lsue_blast2go_fasta_05Oct2017_1535.Renamed.fasta", sep = ""),
             sep = "")
system(cmd)

cmd <- paste("perl ", makerMapFastaIdsPath, " ")



names(fna) <- sub("^contig", "LSUE_G", names(fna))
writeXStringSet(fna, file=paste(sharedPathAn, "Lachnellula_suecica/Lsue_NCBI_files/Lsue_NCBI/contig00047_newName.fna", 
                                sep = ""), append = FALSE, format = "fasta")






```


```{r analysis_and_sequence_directory, cache=TRUE}
analysis   <- "Lachnellula_species_GenomeAn_IonTorrent_2017/"
seqDataDir <- "IonTorrent_data_2017"
# Set the path to the analysis directory:
sharedPathAn <- paste(sharedPath, analysis, sep="")
```

Set paths to Maker and GAG scripts:
```{r}
scriptsPath <- "/home/CFIA-ACIA/girouxeml/prog/scripts_pl/"
makerMapPath <- paste(scriptsPath, "maker_map_ids.pl", sep = "")
makerDatMapPath <- paste(scriptsPath, "map_data_ids.pl", sep = "")
makerMapFastaPath <- paste(scriptsPath, "map_fasta_ids.pl", sep = "")
makerMapGffPath <- paste(scriptsPath, "map_gff_ids.pl", sep = "")
makerFuncGffPath <- paste(scriptsPath, "maker_functional_gff.pl", sep = "")
makerIPRupdatePath <- paste(scriptsPath, "ipr_update_gff.pl", sep = "")
gagPath <- "/home/CFIA-ACIA/girouxeml/prog/gag/genomeannotation-GAG-40ea515/gag.py"
anniePath <- paste("/home/CFIA-ACIA/girouxeml/prog/annie/genomeannotation-annie/annie.py")
```

Set path to Lsue maker and gag output directories:
```{r}
makerOutput <- paste(sharedPathAn, "Lachnellula_suecica/Lsue_NCBI_files/Lsue_NCBI/Maker_Output/", sep = "")
gagOutput <- paste(sharedPathAn, "Lachnellula_suecica/Lsue_NCBI_files/Lsue_NCBI/GAG_output/", sep = "")
lsueNCBIfilesPath <- paste(sharedPathAn, "Lachnellula_suecica/Lsue_NCBI_files/Lsue_NCBI/", sep = "")
gffPath <- paste(lsueNCBIfilesPath, "B2G_project_files/Blast2GO_gff_without_contig.gff", sep = "")
fnaAssPath <- paste(lsueNCBIfilesPath, 
                    "Assembly_Files/FromB2GSupport_Lsue_assembly_EG_2017_mod_removedlLessThan200_contig00022trim.fna", sep = "")
annotPath <- paste(lsueNCBIfilesPath, "B2G_output_For_Maker_analysis/Lsue_blast2go_annot_05Oct2017_1059.annot", sep = "")
annotDesPath <- paste(lsueNCBIfilesPath, "B2G_output_For_Maker_analysis/Lsue_blast2go_annotation_descriptions_05Oct2017_1128.txt", sep = "")
fnaAAPath <- paste(lsueNCBIfilesPath, "B2G_output_For_Maker_analysis/Lsue_blast2go_fasta_05Oct2017_1535.fasta", sep = "")

iprPath <- paste(lsueNCBIfilesPath, "B2G_output_For_Maker_analysis/IPR_Lsue_blast2go_export_05Oct2017_1555.txt", sep = "")
iprBiocluster <- paste(sharedPathAn, "interProScanQsubB2G/Lsue_aa_blast2go.fasta.tsv", sep = "")

# b2gGOannotPath <- paste(makerOutput, "Renamed_Lsue_blast2go_annot_05Oct2017_1059.annot", sep = "")
b2gGOannotPath <- paste(makerOutput, "Lsue_blast2go_annot_30Oct2017.annot", sep = "")

b2gBlastPath <- paste(lsueNCBIfilesPath, "B2G_output_For_Maker_analysis/blast_Lsue_blast2go_export_05Oct2017_1555.txt", sep = "")
gafPath <- paste(lsueNCBIfilesPath, "B2G_output_For_Maker_analysis/Lsue_blast2go_gaf_05Oct2017_1227.txt", sep = "")
b2gBlastGffExportPath <- paste(lsueNCBIfilesPath, "B2G_output_For_Maker_analysis/blast2go_gff_export_05Oct2017_1132.gff", sep = "")
```

Copy files for Maker to Maker_Output directory, to preserve original files, and rename:
```{r}
cpFiles <- c(gffPath, fnaAssPath, fnaAAPath, annotPath, annotDesPath, 
             iprPath, b2gBlastPath, b2gBlastGffExportPath, gafPath)
makerWorkingFiles <- paste(lsueNCBIfilesPath, "Maker_Output/", sep = "")
file.copy(from = cpFiles, to = makerWorkingFiles)
oldNames = list.files(makerWorkingFiles, full.names = TRUE)
oldNames = sapply(strsplit(oldNames, "/"), "[", 12)
newNames = paste0("Renamed_", oldNames)
file.rename(from = file.path(makerWorkingFiles, oldNames), to = file.path(makerWorkingFiles, newNames))
```

Getting data into format we can use: Annie the Genome Annotation Tool:
```{r}
iprBioC <- fread(iprBiocluster)
b2gAnnotGO <- fread(b2gGOannotPath, sep="\t", header=TRUE, fill=TRUE)
```


Rename gene names from gff and other output files for annotations:
Map ID file:
```{r}
prefix = "LSUE_"
justify = 6
abrvGene = "G"
abrvTrans = "T"
gffRenamedPath <- paste(makerOutput, "Renamed_Blast2GO_gff_without_contig.gff", sep = "")
#gffRenamedPath <- paste(makerOutput, "new_Blast2GO_gff_without_contig.gff", sep = "")
mapFilePath <- paste(makerOutput, "id_Lsue.map", sep = "")
modMapFilePath <- paste(makerOutput, "id_Lsue_mod.map", sep = "")
cmd <- paste("perl ", makerMapPath,
             " --prefix ", prefix,
             " --justify ", justify,
             " --abrv_gene ", abrvGene,
             " --abrv_tran ", abrvTrans,
             " ", gffRenamedPath, 
             " > ", mapFilePath, sep = "")

system(cmd)

cmd <- paste(" sed 's/-RA//' ", mapFilePath, " > ", modMapFilePath, sep = "")
system(cmd)
```

Once the names have been changed using maker, in the GFF, they still need to be fixed, so do this 
using the output modified file.
In-place replacement of names with Maker scripts:
GFF and protein fasta files:
```{r}
cmd <- paste("perl ", makerMapGffPath, 
            #" ", modMapFilePath, 
            " ", mapFilePath, 
            " ", gffRenamedPath, 
            sep = "")

system(cmd)

fnaAARenamedPath <- paste(makerOutput, "Renamed_Lsue_blast2go_fasta_05Oct2017_1535.fasta", sep = "")

cmd <- paste("perl ", makerMapFastaPath,
             " ", modMapFilePath,
             " ", fnaAARenamedPath, sep = "")

system(cmd)
gff <- read.table(file = paste(makerOutput, "Renamed_Blast2GO_gff_without_contig.gff", sep = ""), 
                  sep = "\t", header = F, quote = "", comment.char="#", fill = T)

gff_c9 <- subset(gff, select = c(V3,V9))
gff_c9$v1 <- as.character(lapply(strsplit(as.character(gff_c9$V9), split = ";"), "[",1))
gff_c9$v2 <- as.character(lapply(strsplit(as.character(gff_c9$V9), split = "="), "[",2))
gff_c9$v2 <- gsub(";.*", "", gff_c9$v2)
gff_c9$v3 <- as.character(lapply(strsplit(as.character(gff_c9$V9), split = ";"), "[",2))
gff_c9$v4 <- as.character(lapply(strsplit(as.character(gff_c9$V9), split = ";"), "[",3))
gff_c9$v5 <- as.character(lapply(strsplit(as.character(gff_c9$V9), split = ";"), "[",4))
gff_c9$v6 <- as.character(lapply(strsplit(as.character(gff_c9$V9), split = ";"), "[",5))
gff_c9$v7 <- as.character(lapply(strsplit(as.character(gff_c9$V9), split = ";"), "[",6))
unique(gff_c9$v7)

setDT(gff_c9)
gff_c9[ v3 == "Name=", v3 := paste("Name=",v2, sep = "")]
gff_c9[ v4 == "Name=", v4 := paste("Name=",v2, sep = "")]

# if v4 has pattern ^Alias=, then v5 = NA
gff_c9$v5[grepl("Alias", gff_c9$v4)] <- "NA"
# if v5 has pattern ^Alias=, then v6 = NA
gff_c9$v6[grepl("Alias", gff_c9$v5)] <- "NA"
gff_c9$v7[1:10]

gff_c9$v7 <- paste(gff_c9$v1, gff_c9$v3, gff_c9$v4, gff_c9$v5, sep = ";")
gff_c9$v7 <- gsub(";NA", "", gff_c9$v7)
gff_c9$v7 <- gsub("-RA", "", gff_c9$v7)
gff$V9 <- gff_c9$v7
write.table(gff, file = paste(makerOutput, "mod_gff.gff", sep = ""), 
      append = FALSE,
      sep = "\t",
      col.names=FALSE, row.names = FALSE, quote = FALSE)
```


```{r}
# Update the gffRenamedPath to point to the updated, modified gff from this chunk:
gffRenamedPath <- paste(makerOutput, "mod2_gff.gff", sep = "")
```

Stack overflow - add the contig lines to the mod2_gff.gff file, using the terminal:
$ awk 'p!=$1 {print "# Fasta definition line: >" $1;
print "##sequence-region";
p=$1}1' mod_gff.gff > mod2_gff.gff

After this is done, open the gff in gedit, and add the fasta definition lines 1-7 that are in 
the original blast2go gff.

Generate the mRNA and CDS transcripts using the gff and fasta files:
CDS and Coding Sequence files:
```{r}
getAnnoFastaPath <- paste(scriptsPath, "getAnnoFasta.pl", sep = "")

fnaAssRenamedPath <- paste(makerOutput, 
                           "Renamed_FromB2GSupport_Lsue_assembly_EG_2017_mod_removedlLessThan200_contig00022trim.fna",
                           sep = "")

cmd <- paste("perl ", getAnnoFastaPath,
             " ", gffRenamedPath,
             " --seqfile ", fnaAssRenamedPath, 
             #" --protein=on --codingseq=on ", 
             sep = "")

system(cmd)

codingSeqNtFasta <- paste(makerOutput, "cat Renamed_Blast2GO_gff_without_contig.codingseq", sep = "")
```


This is relevant/necessary:
In-place replacement of names with Maker scripts:
Blast files:
```{r}
# Blast file
blastRenamed <- paste(makerOutput, "Renamed_blast_Lsue_blast2go_export_05Oct2017_1555.txt", sep = "")
modBlastrenamed <- paste(makerOutput, "mod_Renamed_blast_Lsue_blast2go_export_05Oct2017_1555.txt", sep = "")

cmd <- paste("perl ", makerDatMapPath,
             " -m ", modMapFilePath,
             " -i ", blastRenamed,
             " > ", modBlastrenamed,
             sep = "")
system(cmd)
```

Good - Keep:
create blast table in format for maker - make it like the uniprot format:
```{r}
library(data.table)
blastTablePath <- paste(makerOutput, "mod_Renamed_blast_Lsue_blast2go_export_05Oct2017_1555.txt", sep = "")

blastTemp <- fread(blastTablePath, sep = "auto", fill = TRUE)

printrows <- blastTemp[1:6]
printrows$`Blast Top Hit Description (HSP)`
gsub("/.*", "", printrows$`Blast Top Hit Description (HSP)`)
blastTemp$`Blast Top Hit Description (HSP)` <- gsub("/.*", "", blastTemp$`Blast Top Hit Description (HSP)`)
blastTemp$gi <- gsub("^([^\\|]*\\|[^\\|]*)\\|.*$", "\\1", blastTemp$`Blast Top Hit Description (HSP)`)
blastTemp$gi <- gsub("^gi\\|", "", blastTemp$gi)
colnames(blastTemp)
colnames(blastTemp)[1] <- "seqName"

topHitDes <- as.data.table(blastTemp$`Blast Top Hit Description (HSP)`)
topHitDes$GN <- as.data.table(blastTemp$`Sequence Description`)
topHitDes$GN <- gsub("---NA---", "", topHitDes$GN)
topHitDes <- topHitDes[!apply(topHitDes == "", 1, all),]

topHitDes
topHitDes$gi <- as.character(lapply(strsplit(as.character(topHitDes$V1), split = "\\|"), "[",1))
topHitDes$geneID <- as.character(lapply(strsplit(as.character(topHitDes$V1), split = "\\|"), "[",2))
topHitDes$ref <- as.character(lapply(strsplit(as.character(topHitDes$V1), split = "\\|"), "[",3))
topHitDes$accession <- as.character(lapply(strsplit(as.character(topHitDes$V1), split = "\\|"), "[",4))
topHitDes$ProteinName <- as.character(lapply(strsplit(as.character(topHitDes$V1), split = "\\|"), "[",5))
topHitDes$OS <- as.character(lapply(strsplit(as.character(topHitDes$ProteinName), split = "\\["), "[",2))
topHitDes$ProteinName <- gsub("\\[.*", "", topHitDes$ProteinName)
topHitDes$OS <- gsub("]", "", topHitDes$OS)

# Make a list of all the gene ids (gi)
giList <- topHitDes$geneID
write(giList, file = paste(makerOutput, "blast_giList.txt", sep = ""), append = FALSE)
```


Good - Keep:

Need to split the lists to use with UniProt due to connection time-out when there are too many queries.
Use command line script:
$ split -l 100 /home/CFIA-ACIA/girouxeml/PIRL_working_directory/Lachnellula_species_GenomeAn_IonTorrent_2017/Lachnellula_suecica/Lsue_NCBI_files/Lsue_NCBI/Maker_Output/blast_giList.txt

To run idmapping through uniprot on all files in directory, using perl script:
$ for i in *; do perl ~/prog/scripts_pl/uniprot_retrieveIDmapping3.pl $i > $i.txt; done
```{r}
library(data.table)
# Copy the gi's to a new column called "EntryName". We will replace the gi's with the entry names
# retrieved from uniprot's retrieve/idmapping perl script using keys and value matches.
topHitDes$EntryName <- topHitDes$geneID

# Add all the results files from the uniprot perl script idmapping to a list:
tempList <- list.files(path = paste(makerOutput, "split_blast_giLists/", sep = ""), pattern = "*.txt$", full.names = TRUE)

# Create an empty csv to which we will add all the results from the uniprot idmapping perl script:
# Create the empty dataframe
DF <- NULL
# read all of the results files into the empty dataframe 'DF':
for (f in tempList){
    dat <- read.csv(f, header = T, sep = "\t", na.strings="", colClasses = "character")
    DF <- rbind(DF, dat)
}

# convert the 'DF' dataframe to a data table so we can do key-value searches and replacements.
DF <- as.data.table(DF)
nrow(DF) # 12246
nrow(topHitDes) # 12697
# There were no matching entry names in uniprot for 451 gi's.

setkey(topHitDes, EntryName)
setkey(DF, From)
# Replace the gi's in the topHitDes$EntryName column with the entry names of the gi's retieved
# by uniprot idmapping
topHitDes[DF, From := To]
topHitDes[, EntryName:=NULL]
setnames(topHitDes, old = c("From"), new = c("EntryName"))

# If you take a look at the DF and the topHitDes, you will see that not all the gi's matched to uniprot 
# entry names. This is normal: http://www.uniprot.org/help/ncbi_mappings
length(topHitDes$EntryName) # 12697

entryNameList <- topHitDes$EntryName
length(entryNameList) # 12697 

# There are several entries that do not have GIs. These must be removed from the list or the 
# looping later on to combine the results will fail.
is.na(entryNameList)
entryNameList <- na.exclude(entryNameList)
length(entryNameList) # 12299

write(entryNameList, file = paste(makerOutput, "blast_UniProt_entryNameList.txt", sep = ""), append = FALSE)
```


Keep - works!

Make a directory in "Maker_Output" called "split_blast_uniprot_entryName2protname" and cd into it.
Run the following in the shell:
$ split -l 100 /home/CFIA-ACIA/girouxeml/PIRL_working_directory/Lachnellula_species_GenomeAn_IonTorrent_2017/Lachnellula_suecica/Lsue_NCBI_files/Lsue_NCBI/Maker_Output/blast_UniProt_entryNameList.txt

Then run the perl script "uniprot_retrieveIDmapping3.pl" to retrieve the ACC_ID and gene names, and 
protein names of the entries in the the split files in the current directory:
$ for i in *; do perl ~/prog/scripts_pl/uniprot_retrieveIDmapping3.pl $i > $i.txt; done

Some of the results will not have anything in uniprot, even though we were able to retrieve entry names.
When I looked at an entry name manually on the uniprot idmapping page, I saw that no uniprot entry name
was recieved, but there was a match to uniparc. It may be that this is what is happing to the others as
well.

We need to remove the results .txt files that are empty, before continuing to loop and collect the 
results or the loop will fail. This means that the script to retrieve the info from uniprot will
need to be repeated later on to retrieve entries that are in uniparc, but not in uniprot.

In the directory where the results are, run the following in the shell:
$ find . -name '*.txt' -size 0 -print0 | xargs -0 rm

Now only files with >0 bytes remain, continue:
```{r}
library(data.table)
# Collect all the results into a list of results:
tempList2 <- list.files(path = paste(makerOutput, "split_blast_uniprot_entryName2protname/", sep = ""), 
                        pattern = "*.txt$", full.names = TRUE)

# Create an empty data table, with the columns I want:
f <- 1
tempDat <- data.table(entry=character(), geneNames=character(), proteinName=character(), entryName=character())

for (f in tempList2){
    dat2 <- read.table(f, header = F, skip=1, sep = "\t", na.strings="", colClasses = "character", 
                       quote = "", fill = T)
    #dat2 <- read.table(tempList2[1], header = F, skip=1, sep = "\t", na.strings="", colClasses = "character", fill = T)
    setnames(dat2, 1:4,c("entry", "geneNames", "proteinName", "entryName"))
    dat2 <- dat2[ , 1:4]
    tempDat <- rbind(tempDat, dat2)
}

# Now add the uniprot protein and gene names to the topHitsDes data table:
# 1. first add the entry name to new column called "proteinName"
topHitDes$proteinName <- topHitDes$EntryName
topHitDes$geneName <- topHitDes$EntryName

# 2. Set up key-value for renaming the protein names:
setkey(topHitDes, proteinName)
setkey(tempDat, entryName)
topHitDes[tempDat, entryName := proteinName]
topHitDes[, proteinName:=NULL]
setnames(topHitDes, old = c("entryName"), new = c("proteinName"))
```


Keep - works!

Because there are uniparc entries, i have to get the entry names and then get the gene names
by querying both uniprot and uniparc separately, then combining the two lists.

Once I have the gene names, in one list, I can get the rest of the info for the protein names
using the gene names.

Need to get info for uniparc entries.
Go from gi -> ACC+ID (Maker_Output/split_blast_giLists)
Add the ACC+ID (entryName) to topHitDes$EntryName
Collect entry names into a list (entryNameList)
make a new dir called "Maker_Output/ split_blast_uniPARC_entryToGeneName"
split the entryNameList into this directory
Run perl idmapping on entryNameList to get gene names for uniparc entries
go from gene names to protein names, etc.

```{r}
library(data.table)
# In the directory where the results are, run the following in the shell:
# $ find . -name '*.txt' -size 0 -print0 | xargs -0 rm
rm(tempDat2, dat2)
# get a list of all the results files:
tempList3 <- list.files(path = paste(makerOutput, "split_blast_uniPARC_entryToGeneName/", sep = ""), 
                        pattern = "*.txt$", full.names = TRUE)

# Create an empty data table, with the columns I want:
f <- 1

tempDat2 <- data.table(entry=character(), geneNames=character(), entryName=character())

for (f in tempList3){
    dat2 <- read.table(f, header = F, skip = 1, sep = "\t", na.strings="", colClasses = "character", 
                       quote = "", fill = T)
    #dat2 <- read.table(tempList2[1], header = F, skip=1, sep = "\t", na.strings="", colClasses = "character", fill = T)
    setnames(dat2, 1:3,c("entry", "geneNames", "entryName"))
    dat2 <- dat2[ , 1:3]
    tempDat2 <- rbind(tempDat2, dat2)
}
tempDat2$geneNames <- gsub(";.*", "", tempDat2$geneNames)

# Rbind
tempDat1 <- tempDat[,c(1,2,4)]
tempDat3 <- rbind(tempDat2, tempDat1)

# Add the gene names to the topHitDes table using key-value assignment:
# 1. Set up key-value for renaming the gene names:
setkey(topHitDes, EntryName)
setkey(tempDat3, entryName)
topHitDes[tempDat3, geneName := geneNames]

# Capture a list of the gene names to run through uniprot idmapping to retrieve protein names,
# which will include uniparc entries.
geneNameList <- topHitDes$geneName

# There are several entries that do not have GIs. These must be removed from the list or the 
# looping later on to combine the results will fail.
length(geneNameList) # 12697
is.na(geneNameList)

geneNameList <- na.exclude(geneNameList)
length(geneNameList) # 11634

write(geneNameList, file = paste(makerOutput, "blast_UniProt_GeneNameList.txt", sep = ""), append = FALSE)
```


Keep - works!

Make a directory called "split_blast_geneNameList_toUniProt_protname". Cd into it and split the 
geneNameList in this directory. Then run the perl script.

$ split -l 100 /home/CFIA-ACIA/girouxeml/PIRL_working_directory/Lachnellula_species_GenomeAn_IonTorrent_2017/Lachnellula_suecica/Lsue_NCBI_files/Lsue_NCBI/Maker_Output/blast_UniProt_GeneNameList.txt

$ for i in *; do perl ~/prog/scripts_pl/uniprot_retrieveIDmapping5.pl $i > $i.txt; done

Remove all files with 0 bytes.
$ find . -name '*.txt' -size 0 -print0 | xargs -0 rm
```{r}
library(data.table)
# get a list of all the results files:
tempList4 <- list.files(path = paste(makerOutput, "split_blast_geneNameList_toUniProt_protname/", sep = ""), 
                        pattern = "*.txt$", full.names = TRUE)

# Create an empty data table, with the columns I want:
f <- 1
tempDat4 <- data.table(entry=character(), entryName=character(), proteinName=character(), 
                       geneName=character(), proteinEx=character(), seqVers=character())

for (f in tempList4){
    dat4 <- read.table(f, header = T, sep = "\t", na.strings="", colClasses = "character", 
                       quote = "", fill = F)
    cols <- grep("Entry|Entry.name|Protein.names|your|Protein.existence|Version", names(dat4), value = TRUE)
    datTemp <- subset(dat4, select = c(cols))
    colnames(datTemp)
    colnames(datTemp) <- c("entry", "entryName", "proteinName", "proteinEx", "seqVers", "geneName")
    tempDat4 <- rbind(tempDat4, datTemp)
}
colnames(tempDat4)
length(tempDat4$entry)
unique(tempDat4$entryName)
unique(topHitDes$EntryName)
cleanedtempDat4 <- tempDat4[tempDat4$entryName %in% topHitDes$EntryName,]

# Now add the uniprot protein and gene names to the topHitsDes data table:
# 1. Set up key-value for renaming the protein names:
setkey(topHitDes, EntryName)
setkey(cleanedtempDat4, entryName)
topHitDes[cleanedtempDat4, proteinName := proteinName]
topHitDes[cleanedtempDat4, entry := entry]
topHitDes[cleanedtempDat4, proteinEx := proteinEx]
topHitDes[cleanedtempDat4, seqVers := seqVers]

topHitDes$proteinEx <- gsub("Predicted", "4", topHitDes$proteinEx)
topHitDes$proteinEx <- gsub("Inferred from homology", "3", topHitDes$proteinEx)
topHitDes$proteinEx <- gsub("Evidence at protein level", "1", topHitDes$proteinEx)
topHitDes$proteinEx <- gsub("Evidence at transcript level", "2", topHitDes$proteinEx)
unique(topHitDes$proteinEx)
```

Keep - works!
The following will complete the uniprot headers for those entries that successfully returned 
protein names. 
```{r}
library(data.table)
topHitDes$uniProtHeader <- paste(">sp|",topHitDes$entry, "|", topHitDes$EntryName, " ", topHitDes$proteinName,
                                 " OS=", topHitDes$OS, " GN=", topHitDes$geneName,
                                 " PE=", topHitDes$proteinEx, " SV=", topHitDes$seqVers, sep = "")
# Collect those with protein names into a new set:
completeTopHits <- topHitDes[!is.na(topHitDes$proteinName),]
write(completeTopHits$uniProtHeader, 
      file = paste(makerOutput, "sp_completeUniprotHits_db.fa", sep = ""), 
      append = FALSE)
# Prepare the matching blast output for the uniprot hits:
blastSeqNameGI <- subset(blastTemp, select = c("seqName", "gi"))

# check there are common values:
common <- intersect(blastSeqNameGI$gi, completeTopHits$geneID)

# 1. Set up key-value for assigning the sequence names:
completeTopHits$seqName <- NULL
setkey(completeTopHits, geneID)
setkey(blastSeqNameGI, gi)
completeTopHits[blastSeqNameGI, seqName := seqName]

# Add a column that replicates the blast output format:
completeTopHits$blastOutComplete <- paste("sp|", completeTopHits$entry, "|", completeTopHits$EntryName, sep = "")
blastTab <- subset(completeTopHits, select = c("seqName", "blastOutComplete"))

# Replicate the blast output:
write.table(blastTab, 
      file = paste(makerOutput, "sp_completeBlastHits_db.txt", sep = ""), 
      append = FALSE,
      sep = "\t",
      col.names=FALSE, row.names = FALSE, quote = FALSE)
```

Keep - works!
Test to see if Maker works on our replicated uniprot database and blast output:
```{r}
uniprotDB <- paste(makerOutput, "sp_completeUniprotHits_db.fa", sep = "")
blastUniprot <- paste(makerOutput, "sp_completeBlastHits_db.txt", sep = "")

cmd <- paste("perl ", makerFuncGffPath,
             " ", uniprotDB,
             " ", blastUniprot, 
             " ", gffRenamedPath,
             " > ", gffRenamedFuncPath, sep = "")

system(cmd)
```

We have incomplete headers for those that did not retrieve uniprot entry or entry names, and by default
aslo protein names. We need to fix these as well.
```{r}
library(data.table)
# Collect those with no protein names into a new set:
restTopHits <- topHitDes[is.na(topHitDes$proteinName),]

# Those with entry names, but no protein name (i.e., uniPARC matches):
uniParcHits <- restTopHits[!is.na(restTopHits$EntryName)]

# Capture a list of the gene names to run through uniprot idmapping to retrieve protein names,
# which will include uniparc entries.
entryNameList <- uniParcHits$EntryName

# There are several entries that do not have GIs. These must be removed from the list or the 
# looping later on to combine the results will fail.
length(entryNameList) # 3865
is.na(entryNameList)

entryNameList <- na.exclude(entryNameList)
length(entryNameList) # 3865

write(entryNameList, file = paste(makerOutput, "blast_UniParc_entryNameList.txt", sep = ""), append = FALSE)
```
Keep - works!

Make a directory in "Maker_Output" called "split_blast_UniParc_UID" and cd into it.
Run the following in the shell:
$ split -l 100 /home/CFIA-ACIA/girouxeml/PIRL_working_directory/Lachnellula_species_GenomeAn_IonTorrent_2017/Lachnellula_suecica/Lsue_NCBI_files/Lsue_NCBI/Maker_Output/blast_UniParc_entryNameList.txt

Then run the perl script "uniprot_retrieveIDmapping6.pl" to retrieve the ACC_ID and gene names, and 
protein names of the entries in the the split files in the current directory:
$ for i in *; do perl ~/prog/scripts_pl/uniprot_retrieveIDmapping6.pl $i > $i.txt; done

To remove empty files:
In the directory where the results are, run the following in the shell:
$ find . -name '*.txt' -size 0 -print0 | xargs -0 rm

Now only files with >0 bytes remain, continue:
```{r}
library(data.table)
# get a list of all the results files:
tempList5 <- list.files(path = paste(makerOutput, "split_blast_UniParc_UID/", sep = ""), 
                        pattern = "*.txt$", full.names = TRUE)

# Create an empty data table, with the columns I want:
f <- 1
tempDat5 <- data.table(entry=character(), entryName=character(), geneName=character(), 
                       proteinName=character(), proteinEx=character(), seqVers=character(),
                       uniparcID=character())

for (f in tempList5){
    dat5 <- read.table(f, header = T, sep = "\t", na.strings="", colClasses = "character", 
                       quote = "", fill = F)
    cols <- grep("Entry|Entry.name|Gene|Protein.names|Protein.existence|Version|your", names(dat5), value = TRUE)
    datTemp2 <- subset(dat5, select = c(cols))
    colnames(datTemp2)
    colnames(datTemp2) <- c("entry", "entryName", "geneName", "proteinName", "proteinEx", "seqVers", "uniparcID")
    tempDat5 <- rbind(tempDat5, datTemp2)
}
colnames(tempDat5)
length(tempDat5$uniparcID) # 3762
unique(tempDat5$uniparcID)
unique(topHitDes$EntryName)

# There are more than one row for some of the uniparc ID's. We must remove all but the first row for each UID.
cleanedtempDat5 <- unique(tempDat5, by =c("uniparcID"))
cleanedtempDat5 <- cleanedtempDat5[cleanedtempDat5$uniparcID %in% topHitDes$EntryName,]

is.data.table(cleanedtempDat5)
is.data.table(uniParcHits)
summary(cleanedtempDat5)
```

Note: After doing the retrieval using the uniParc IDs (UPI), there are still some entries in the uniParcList
that are not returning information. I looked at these by taking a sample and querying them on the UniProt Retrieve/ID mapping
page, and these show up as having UniProtKB entries that are obsolete. This makes up 104 of those searched
from the uniparc list.
```{r}
# Clean up any set keys:
setkey(topHitDes, NULL)
setkey(cleanedtempDat5, NULL)
setkey(tempDat5, NULL)
setkey(uniParcHits, NULL)

# Can't add values to column that already exists, so working with the dt uniParcHits that has the 
# subset of rows for uniparc entries from the topHitsDes dt, remove the proteinName column:
uniParcHits$proteinName <- NULL
uniParcHits$proteinEx <- NULL
uniParcHits$geneName <- NULL
uniParcHits$seqVers <- NULL
uniParcHits$uniProtHeader <- NULL
uniParcHits$entry <- NULL
uniParcHits$entryName2 <- NULL
uniParcHits$uniparcID <- NULL
names(uniParcHits)

# Now add the uniparc protein and gene names to the uniParcHits data table:
# 1. Set up key-value for renaming the protein names:
setkey(uniParcHits, EntryName)
setkey(cleanedtempDat5, uniparcID)

uniParcHits[cleanedtempDat5, uniparcID := uniparcID]
uniParcHits[cleanedtempDat5, EntryName2 := entryName] # entry name given from uniParc UPI ID.
uniParcHits[cleanedtempDat5, geneName := geneName]
uniParcHits[cleanedtempDat5, proteinName := proteinName]
uniParcHits[cleanedtempDat5, entry := entry]
uniParcHits[cleanedtempDat5, proteinEx := proteinEx]
uniParcHits[cleanedtempDat5, seqVers := seqVers]

# Replace protein existence descriptors with corresponding numeric values:
uniParcHits$proteinEx <- gsub("Predicted", "4", uniParcHits$proteinEx)
uniParcHits$proteinEx <- gsub("Inferred from homology", "3", uniParcHits$proteinEx)
uniParcHits$proteinEx <- gsub("Evidence at protein level", "1", uniParcHits$proteinEx)
uniParcHits$proteinEx <- gsub("Evidence at transcript level", "2", uniParcHits$proteinEx)
unique(uniParcHits$proteinEx)
```

The following will complete the uniparc headers in uniprot format for those entries that successfully returned 
protein names. 
```{r}
library(data.table)
uniParcHits$uniProtHeader <- paste(">sp|",uniParcHits$entry, "|", uniParcHits$EntryName2, " ", uniParcHits$proteinName,
                                 " OS=", uniParcHits$OS, " GN=", uniParcHits$geneName,
                                 " PE=", uniParcHits$proteinEx, " SV=", uniParcHits$seqVers, sep = "")

# Collect those with protein names into a new set:
completeTopHitsUniParc <- uniParcHits[!is.na(uniParcHits$proteinName),]
write(completeTopHitsUniParc$uniProtHeader, 
      file = paste(makerOutput, "sp_completeUniParcHits_db.fa", sep = ""), 
      append = FALSE)


# Prepare the matching blast output for the uniprot hits:
blastSeqNameGI <- subset(blastTemp, select = c("seqName", "gi"))

# check there are common values:
common2 <- intersect(blastSeqNameGI$gi, completeTopHitsUniParc$geneID)

# 1. Set up key-value for assigning the sequence names:
#completeTopHitsUniParc$seqName <- NULL
setkey(completeTopHitsUniParc, geneID)
setkey(blastSeqNameGI, gi)
completeTopHitsUniParc[blastSeqNameGI, seqName := seqName]

# Add a column that replicates the blast output format:
completeTopHitsUniParc$blastOutComplete <- paste("sp|", completeTopHitsUniParc$entry, "|", completeTopHitsUniParc$EntryName2, sep = "")
blastTab2 <- subset(completeTopHitsUniParc, select = c("seqName", "blastOutComplete"))

# Replicate the blast output:
write.table(blastTab2, 
      file = paste(makerOutput, "sp_completeBlastHitsUniParc_db.txt", sep = ""), 
      append = FALSE,
      sep = "\t",
      col.names=FALSE, row.names = FALSE, quote = FALSE)
```

Make a combined blast database text file, and uniprot header database fasta file:
```{r}
blastDb_union <- rbind(blastTab, blastTab2)
blastFastadb_union <- union(completeTopHits$uniProtHeader, completeTopHitsUniParc$uniProtHeader)

# Collect those with protein names into a new set:
write(blastFastadb_union, 
      file = paste(makerOutput, "sp_Union_complete_db.fa", sep = ""), 
      append = FALSE)

# Replicate the blast output:
write.table(blastDb_union, 
      file = paste(makerOutput, "sp_Union_completeBlast_db.txt", sep = ""), 
      append = FALSE,
      sep = "\t",
      col.names=FALSE, row.names = FALSE, quote = FALSE)
```

Keep - works!
Test to see if Maker works on our replicated uniparc/uniprot database and blast output:
Note, I'm adding to the already modified gff, from the first update (prior to repeating only uniparc missing entries)
```{r}
uniprot_uniparc_union <- paste(makerOutput, "sp_Union_complete_db.fa", sep = "")
blast_union_uniprot_uniparc <- paste(makerOutput, "sp_Union_completeBlast_db.txt", sep = "")

gffRenamedFuncPath_union <- paste(makerOutput, "func_Union_Renamed_Blast2GO_gff_without_contig.gff", sep = "")

cmd <- paste("perl ", makerFuncGffPath,
             " ", uniprot_uniparc_union,
             " ", blast_union_uniprot_uniparc, 
             " ", gffRenamedPath,
             " > ", gffRenamedFuncPath_union, sep = "")

system(cmd)
```

See what remains after running uniprot and uniparc id mapping. I still see some that appear to be mappable
from uniprot to uniprot entries, i.e., A0A0F6N2A4_9EURO
```{r}
# Collect those with no protein names into a new set:
restTopHitsUniPa <- uniParcHits[is.na(uniParcHits$uniparcID),]
restTopHitsLast <- restTopHits[is.na(restTopHits$EntryName),]

# Start with retrieving/id mapping the rest from the uniParcTopHits:
# Capture a list of their gene IDs (gi) - seems to work on web query.
giList2 <- restTopHitsUniPa$geneID
length(giList2) # 104
is.na(giList2) # none are NA - good

write(giList2, file = paste(makerOutput, "blast_UniParcRest_giList.txt", sep = ""), 
      append = FALSE)

# Also write a list to search by the EntryName, maybe get better hit with reformatting the perl
# script for querying
entryNameListUniParcRest <- restTopHitsUniPa$EntryName
write(entryNameListUniParcRest, 
      file = paste(makerOutput, "blast_UniParcRest_EntryNameList.txt", sep = ""), 
      append = FALSE)
```

Make a directory to put the results of the id mapping for the rest of the gi's from the unmapped Uniparc results:
```{r}
dir.create(paste(makerOutput, "blast_UniParcRest_giList", sep = ""), showWarnings = TRUE, recursive = FALSE)
file.copy(paste(makerOutput, "blast_UniParcRest_giList.txt", sep = ""), 
          paste(makerOutput, "blast_UniParcRest_giList/", sep = ""))
file.copy(paste(makerOutput, "blast_UniParcRest_EntryNameList.txt", sep = ""), 
          paste(makerOutput, "blast_UniParcRest_giList/", sep = ""))
```
There is no need to split this file, since it only has 104 entries.
Continue by running the perl script:
cd into the new directory and run:
$ perl  ~/prog/scripts_pl/uniprot_retrieveIDmapping7.pl blast_UniParcRest_giList.txt > blast_UniParcRest_giList.txt.txt

The results came back for 17 of the 104 entries:
```{r}
# get a list of all the results files:
tempList5 <- list.files(path = paste(makerOutput, "split_blast_UniParc_UID/", sep = ""), 
                        pattern = "*.txt$", full.names = TRUE)

# Create an empty data table, with the columns I want:
temp6 <- read.table(paste(makerOutput, "blast_UniParcRest_giList/blast_UniParcRest_giList.txt.txt", sep = ""),
                    header = T, sep = "\t", na.strings="", colClasses = "character", quote = "", fill = F)

cols <- grep("Entry|Entry.name|Gene|Protein.names|Protein.existence|Version|your", names(temp6), value = TRUE)
tempDat6 <- subset(temp6, select = c(cols))

# Note all the entries recieved do not have gene names assigned. Omit these and move on to adding
# interpro go annotations.
```

Continue by adding the putative putative gene functions to the fasta files, the ones generated for transcripts
and the ones generated for proteins:
```{r}
exonMakerpath <- paste(makerOutput, "Renamed_Blast2GO_gff_without_contig.cdsexons", sep = "")
transcrMaker <- paste(makerOutput, "Renamed_Blast2GO_gff_without_contig.codingseq", sep = "")
protFna <- paste(makerOutput, "Renamed_Lsue_blast2go_fasta_05Oct2017_1535.fasta", sep = "")
makerFuncFastaPath <- "/home/CFIA-ACIA/girouxeml/prog/maker/maker/src/bin/maker_functional_fasta"

cmd <- paste("perl ", makerFuncFastaPath,
             " ", uniprot_uniparc_union,
             " ", blast_union_uniprot_uniparc,
             " ", protFna,
             " > ", paste(makerOutput, "Renamed_FromB2GSupport_Lsue_assembly_putativeFunction.fna", sep = ""),
             sep = "")
system(cmd)

# Note - this is not correct - there are names already in this file and they become disordered. 
# Look back at the original protein output and see how they are different and what is being updated.
# Also, the blast2go output had names included, much better than my attempts to manually fix them. See
# if it matters if I leave them as is, or what needs to be changed is they have to be made congruent
# between gff and fna files.

# The process is to rename the transcripts and protein fasta files with the uniprot/uniparc headers I made.
```

Below the interPro GO id's and terms are added to the gff that we've just added the top blast hits to:
```{r}
library(data.table)
annotPath2 <- paste(makerOutput, "Lsue_blast2go_annot_30Oct2017.annot", sep = "")
annotTbl2 <- fread(annotPath2, sep = "\t", header=TRUE, fill=TRUE)
# unique(annotTbl2$`Enzyme Codes`)

tblPath <- paste(makerOutput, "blast2go_export_some_30Oct2017.txt", sep = "")
tbl <- fread(tblPath, sep = "\t", header=TRUE, fill=TRUE)
colnames(tbl)
tbl$`Mapping Taxa ID`[1]
#unique(tbl$`InterPro Signatures`)
tblColNames <- c("seqName", "seqLen", "interProName", "interProSignatures",
                 "interProGOid", "interProGOterm", "interProGOcategory", "interProAccn",
                 "interProType", "mappingGeneName", "mappingTaxaID", "mappingXref", 
                 "mappingXrefDB", "mappingGOid", "mappingGOterm", "mappingGOcategory",
                 "annotGOcount", "annotGOid", "annotGOterm", "annotGOcategory")
colnames(tbl) <- tblColNames

# Collect only the rows for which there are values in the column "interProAccn"
# replace the term noIPR in interProAccn with nothing:
tbl$interProAccn <- gsub(pattern = "noIPR", "", tbl$interProAccn)
tblSubset <- tbl[tbl$interProAccn != "",]
tblSubset$annotGOcount <- NULL

# remove tbl from the global environment now that we have the subset of it that we want, 
# becuase tbl is a huge file, and may contribute to system crashes.
rm(tbl)

# Split up the interPro ethod and database id's correctly:
tblSubset$interProSignatures1 <- sapply(strsplit(as.character(tblSubset$interProSignatures),' '), "[", 1)
tblSubset$interProSignatures2 <- sapply(strsplit(as.character(tblSubset$interProSignatures),' '), "[", 2)
tblSubset$interProSignatures2 <- gsub("[[:punct:]]", "", tblSubset$interProSignatures2)


# Split the table into two, one set that has annotGOid values, the other that doesn't but has interPro GO ids
# Set with annotation GO ids:
tblSubAnnotGoids <- tblSubset[tblSubset$annotGOid !="",]
tblAnnieA <- tblSubAnnotGoids[,c("seqName", 
                  "seqName",
                  "interProGOid",
                  "interProSignatures2", 
                  "interProSignatures1",
                  "interProName", 
                  "interProType",
                  "interProGOterm",
                  "mappingGeneName",
                  "mappingXref",
                  "mappingXrefDB",
                  "interProAccn",
                  "annotGOterm",
                  "annotGOid")]
# rename the two columns that differ between tblAnnieA and tblAnnieB:
setnames(tblAnnieA, "interProGOid", "interPro_annotGODid")
setnames(tblAnnieA, "annotGOid", "annot_interProGOid")

# Set without annotation GO ids, but with interpro GO is:
tblSubIPRGoids <- tblSubset[tblSubset$annotGOid =="",]
tblAnnieB <- tblSubIPRGoids[,c("seqName", 
                  "seqName",
                  "annotGOid",
                  "interProSignatures2", 
                  "interProSignatures1",
                  "interProName",
                  "interProType",
                  "interProGOterm", 
                  "mappingGeneName",
                  "mappingXref",
                  "mappingXrefDB",
                  "interProAccn",
                  "annotGOterm",
                  "interProGOid")]

# rename the two columns that differ between tblAnnieA and tblAnnieB:
setnames(tblAnnieB, "annotGOid", "interPro_annotGODid")
setnames(tblAnnieB, "interProGOid", "annot_interProGOid")


# rbind the tables, sort by seqName 
tblAnnie <- rbind(tblAnnieA, tblAnnieB)

# Sort the takble so it's ordered by seqName:
setkey(tblAnnie, "seqName")

# remove tblSubIPRGoids, tblSubset, tblAnnieA and tblAnnieB from the global environment
# as these are huge files and may contribute to system crashes:
rm(tblSubAnnotGoids, tblSubset, tblAnnieA, tblAnnieB)

# write the table as tab delimited text file:
write.table(tblAnnie, 
      file = paste(makerOutput, "annieIPR_table.txt", sep = ""),
      sep = "\t",
      quote = FALSE,
      col.names = FALSE,
      row.names = FALSE,
      append = FALSE)

pathAnnieIPRtbl <- paste(makerOutput, "annieIPR_table.txt", sep = "")
# remove tblAnnie from the global environment. If we want it, we can read it back in since
# we've written it to file:
rm(tblAnnie)
```

Run annie.py in order to get the 3-column table for dbxref, IPR and GO id:
$ python3 ~/prog/annie/genomeannotation-annie/annie.py -ipr annieIPR_table.txt

The keys are off - see issue:
https://github.com/genomeannotation/annie/issues/6 

Need to fix this, since there is no fix or new release:
```{r}
annieTbltoFix <- fread(file = paste(makerOutput, "annie_output.tsv", sep = ""), sep = "\t",
                       header = FALSE)
                       
annieTbltoFix$V2[grepl("GO", annieTbltoFix$V3)] <- "GO"
annieTbltoFix$V2[grepl("InterPro", annieTbltoFix$V3)] <- "InterPro"
annieTbltoFix$V3 <- gsub("InterPro:", "", annieTbltoFix$V3 )

# write the table as tab delimited text file:
write.table(annieTbltoFix, 
      file = paste(makerOutput, "annie_fixed_output.txt", sep = ""),
      sep = "\t",
      quote = FALSE,
      col.names = FALSE,
      row.names = FALSE,
      append = FALSE)

# record the path of the new Annie table:
anniePathFinal <- paste(makerOutput, "annie_fixed_output.txt", sep = "")

# remove the annieTbltoFix since we've written it to file and don't need to keep this
# in the global environment:
rm(annieTbltoFix)
```

Need to update the contig names to the new ones created with Maker, in our Annie Table:
In-place replacement of names with Maker scripts:
```{r}
# I'm choosing to add the GO id's and interPro terms using Maker instead of GAG using Annie's 3-column
# table because I noticed that using GAG creates a gff3/gtf2 hybrid that has quotes, and terminal semilcolons
# and is not consistent with the Maker structure I've been using. It doesn't necessarily mean that is wrong,
# but I think it may create more difficulties.
modAnnieIPRfullRenamed <- paste(makerOutput, "renamed_annieIPR_table.txt", sep = "")
cmd <- paste("perl ", makerDatMapPath,
             " -m ", modMapFilePath,
             " -i ", pathAnnieIPRtbl,
             " > ", modAnnieIPRfullRenamed,
             sep = "")
system(cmd)


# below is the file we need to rename if we are to use the 3-column Annie table
# to add interPro annotations. Note, I used the above with Maker instead.:

# First lets set a path and name for the updated output file we're making of our Annie table:
# modAnnieRenamed <- paste(makerOutput, "renamed_annie_fixed_output.txt", sep = "")
# cmd <- paste("perl ", makerDatMapPath,
#              " -m ", modMapFilePath,
#              " -i ", anniePathFinal,
#              " > ", modAnnieRenamed,
#              sep = "")
# system(cmd)
```


Finally we will add protein domain information to the final annotations using a report from InterProScan. This is down using the following scripts:
ipr_update_gff - adds searchable tags to the gene and mRNA features in the GFF3 files
iprscan2gff3 - adds physical viewable features for daomains that can be displayed in JBrowse, Gbrowse, and Web Apollo.
ipr_update_gff hsap_contig.renamed.putative_function.gff output.renamed.iprscan > hsap_contig.renamed.putative_function.domain_added.gff
iprscan2gff3 output.renamed.iprscan hsap_contig.renamed.gff > visible_iprscan_domains.gff 

```{r}
iprUpdateMaker <- "/home/CFIA-ACIA/girouxeml/prog/maker/maker/src/bin/ipr_update_gff"
gffRenamedFuncDomain <- paste(makerOutput, "func_Union_Domain_Renamed_Blast2GO_gff_without_contig.gff", sep = "")

cmd <- paste("perl ", iprUpdateMaker,
             " ", gffRenamedFuncPath_union,
             " ", modAnnieIPRfullRenamed,
             " > ", gffRenamedFuncDomain,
             sep = "")
system(cmd)



```


##########################################################################################
Extra unwanted:


InterProScan GO annotations:
Rename the contig names in the interPro output for
In-place replacement of names with Maker scripts:
```{r}
iprRenamed <- paste(makerOutput, "Renamed_IPR_Lsue_blast2go_export_05Oct2017_1555.txt", sep = "")
modIPRrenamed <- paste(makerOutput, "mod_Renamed_IPR_Lsue_blast2go_export_05Oct2017_1555.txt", sep = "")

# To see what the files look like:
interPro <- fread(iprPath)
modMap <- fread(modMapFilePath)

cmd <- paste("perl ", makerDatMapPath,
             " -m ", modMapFilePath,
             " -i ", iprRenamed,
             " > ", modIPRrenamed,
             sep = "")
system(cmd)

# Take a look now:
modInterPro <- fread(modIPRrenamed)

# Remove the files to conserve space:
rm(interPro, modMap)
```

Add annotations from IPR and blast to gff file using maker:
This isn't working for some reason, the data from the blast output from blast2go is not formatted 
to work with this properly. It needs to be in UniProt or UniRef format. 

Skip this chunk and see if I can convert the blast data in the xml file to a 
table format, and see if this would work better.
```{r}
gffRenamedFuncPath <- paste(makerOutput, "func_Renamed_Blast2GO_gff_without_contig.gff", sep = "")
uniProtDBPath <- "/home/CFIA-ACIA/girouxeml/Databases/uniprot/uniprot_sprot.fasta"
gffIPRupdate <- paste(makerOutput, "IPR_Renamed_Blast2GO_gff_without_contig.gff", sep = "")

cmd <- paste("perl ", makerIPRupdatePath,
             " ", gffRenamedPath,
             " ", iprRenamed, 
             " > ", gffIPRupdate, sep = "")

system(cmd)

```


Set paths to Maker and GAG scripts:
```{r}
scriptsPath <- "/home/CFIA-ACIA/girouxeml/prog/scripts_pl/"
makerMapPath <- paste(scriptsPath, "maker_map_ids.pl", sep = "")
makerDatMapPath <- paste(scriptsPath, "map_data_ids.pl", sep = "")
makerMapFastaPath <- paste(scriptsPath, "map_fasta_ids.pl", sep = "")
makerMapGffPath <- paste(scriptsPath, "map_gff_ids.pl", sep = "")
makerFuncGffPath <- paste(scriptsPath, "maker_functional_gff.pl", sep = "")
makerIPRupdatePath <- paste(scriptsPath, "ipr_update_gff.pl", sep = "")
gagPath <- "/home/CFIA-ACIA/girouxeml/prog/gag/genomeannotation-GAG-40ea515/gag.py"
anniePath <- paste("/home/CFIA-ACIA/girouxeml/prog/annie/genomeannotation-annie/annie.py")
```

Set path to Lsue maker and gag output directories:
```{r}
makerOutput <- paste(sharedPathAn, "Lachnellula_suecica/Lsue_NCBI_files/Lsue_NCBI/Maker_Output/", sep = "")
gagOutput <- paste(sharedPathAn, "Lachnellula_suecica/Lsue_NCBI_files/Lsue_NCBI/GAG_output/", sep = "")
lsueNCBIfilesPath <- paste(sharedPathAn, "Lachnellula_suecica/Lsue_NCBI_files/Lsue_NCBI/", sep = "")
gffPath <- paste(lsueNCBIfilesPath, "B2G_project_files/Blast2GO_gff_without_contig.gff", sep = "")
fnaAssPath <- paste(lsueNCBIfilesPath, "Assembly_Files/FromB2GSupport_Lsue_assembly_EG_2017_mod_removedlLessThan200_contig00022trim.fna", sep = "")
annotPath <- paste(lsueNCBIfilesPath, "B2G_output_For_Maker_analysis/Lsue_blast2go_annot_05Oct2017_1059.annot", sep = "")
annotDesPath <- paste(lsueNCBIfilesPath, "B2G_output_For_Maker_analysis/Lsue_blast2go_annotation_descriptions_05Oct2017_1128.txt", sep = "")
fnaAAPath <- paste(lsueNCBIfilesPath, "B2G_output_For_Maker_analysis/Lsue_blast2go_fasta_05Oct2017_1535.fasta", sep = "")

iprPath <- paste(lsueNCBIfilesPath, "B2G_output_For_Maker_analysis/IPR_Lsue_blast2go_export_05Oct2017_1555.txt", sep = "")
iprBiocluster <- paste(sharedPathAn, "interProScanQsubB2G/Lsue_aa_blast2go.fasta.tsv", sep = "")

# b2gGOannotPath <- paste(makerOutput, "Renamed_Lsue_blast2go_annot_05Oct2017_1059.annot", sep = "")
b2gGOannotPath <- paste(makerOutput, "Lsue_blast2go_annot_30Oct2017.annot", sep = "")

b2gBlastPath <- paste(lsueNCBIfilesPath, "B2G_output_For_Maker_analysis/blast_Lsue_blast2go_export_05Oct2017_1555.txt", sep = "")
gafPath <- paste(lsueNCBIfilesPath, "B2G_output_For_Maker_analysis/Lsue_blast2go_gaf_05Oct2017_1227.txt", sep = "")
b2gBlastGffExportPath <- paste(lsueNCBIfilesPath, "B2G_output_For_Maker_analysis/blast2go_gff_export_05Oct2017_1132.gff", sep = "")
```

Copy files for Maker to Maker_Output directory, to preserve original files, and rename:
```{r}
cpFiles <- c(gffPath, fnaAssPath, fnaAAPath, annotPath, annotDesPath, 
             iprPath, b2gBlastPath, b2gBlastGffExportPath, gafPath)
makerWorkingFiles <- paste(lsueNCBIfilesPath, "Maker_Output/", sep = "")
file.copy(from = cpFiles, to = makerWorkingFiles)
oldNames = list.files(makerWorkingFiles, full.names = TRUE)
oldNames = sapply(strsplit(oldNames, "/"), "[", 12)
newNames = paste0("Renamed_", oldNames)
file.rename(from = file.path(makerWorkingFiles, oldNames), to = file.path(makerWorkingFiles, newNames))
```

Getting data into format we can use: Annie the Genome Annotation Tool:
```{r}
iprBioC <- fread(iprBiocluster)
b2gAnnotGO <- fread(b2gGOannotPath, sep="\t", header=TRUE, fill=TRUE)
```


Rename gene names from gff and other output files for annotations:
Map ID file:
```{r}
prefix = "LSUE_"
justify = 6
abrvGene = "G"
abrvTrans = "T"
gffRenamedPath <- paste(makerOutput, "Renamed_Blast2GO_gff_without_contig.gff", sep = "")
#gffRenamedPath <- paste(makerOutput, "new_Blast2GO_gff_without_contig.gff", sep = "")
mapFilePath <- paste(makerOutput, "id_Lsue.map", sep = "")
modMapFilePath <- paste(makerOutput, "id_Lsue_mod.map", sep = "")
cmd <- paste("perl ", makerMapPath,
             " --prefix ", prefix,
             " --justify ", justify,
             " --abrv_gene ", abrvGene,
             " --abrv_tran ", abrvTrans,
             " ", gffRenamedPath, 
             " > ", mapFilePath, sep = "")

system(cmd)

cmd <- paste(" sed 's/-RA//' ", mapFilePath, " > ", modMapFilePath, sep = "")
system(cmd)
```

Once the names have been changed using maker, in the GFF, they still need to be fixed, so do this 
using the output modified file.
In-place replacement of names with Maker scripts:
GFF and protein fasta files:
```{r}
cmd <- paste("perl ", makerMapGffPath, 
            #" ", modMapFilePath, 
            " ", mapFilePath, 
            " ", gffRenamedPath, 
            sep = "")

system(cmd)

fnaAARenamedPath <- paste(makerOutput, "Renamed_Lsue_blast2go_fasta_05Oct2017_1535.fasta", sep = "")

cmd <- paste("perl ", makerMapFastaPath,
             " ", modMapFilePath,
             " ", fnaAARenamedPath, sep = "")

system(cmd)
gff <- read.table(file = paste(makerOutput, "Renamed_Blast2GO_gff_without_contig.gff", sep = ""), 
                  sep = "\t", header = F, quote = "", comment.char="#", fill = T)

gff_c9 <- subset(gff, select = c(V9))
gff_c9$v1 <- as.character(lapply(strsplit(as.character(gff_c9$V9), split = ";"), "[",1))
gff_c9$v2 <- as.character(lapply(strsplit(as.character(gff_c9$V9), split = "="), "[",2))
gff_c9$v2 <- gsub(";.*", "", gff_c9$v2)
gff_c9$v3 <- as.character(lapply(strsplit(as.character(gff_c9$V9), split = ";"), "[",2))
gff_c9$v4 <- as.character(lapply(strsplit(as.character(gff_c9$V9), split = ";"), "[",3))
gff_c9$v5 <- as.character(lapply(strsplit(as.character(gff_c9$V9), split = ";"), "[",4))
gff_c9$v6 <- as.character(lapply(strsplit(as.character(gff_c9$V9), split = ";"), "[",5))
gff_c9$v7 <- as.character(lapply(strsplit(as.character(gff_c9$V9), split = ";"), "[",6))
unique(gff_c9$v7)

setDT(gff_c9)
gff_c9[ v3 == "Name=", v3 := paste("Name=",v2, sep = "")]
gff_c9[ v4 == "Name=", v4 := paste("Name=",v2, sep = "")]

# gff_c9$v3 <- gsub("^Name=", paste("Name=", gff_c9$v2, sep = ""), gff_c9$v3)
# gff_c9$v4 <- gsub("^Name=", paste("Name=", gff_c9$v2, sep = ""), gff_c9$v4)

# if v4 has pattern ^Alias=, then v5 = NA
gff_c9$v5[grepl("Alias", gff_c9$v4)] <- "NA"
# if v5 has pattern ^Alias=, then v6 = NA
gff_c9$v6[grepl("Alias", gff_c9$v5)] <- "NA"
gff_c9$v7[1:10]

#unique(gff_c9$v4)
#gff_c9$v7 <- paste(gff_c9$v1, gff_c9$v3, gff_c9$v4, gff_c9$v5, gff_c9$v6, sep = ";")

gff_c9$v7 <- paste(gff_c9$v1, gff_c9$v3, gff_c9$v4, gff_c9$v5, sep = ";")
gff_c9$v7 <- gsub(";NA", "", gff_c9$v7)
gff_c9$v7 <- gsub("-RA", "", gff_c9$v7)
gff$V9 <- gff_c9$v7
write.table(gff, file = paste(makerOutput, "mod_gff.gff", sep = ""), 
      append = FALSE,
      sep = "\t",
      col.names=FALSE, row.names = FALSE, quote = FALSE)
```


```{r}
# Update the gffRenamedPath to point to the updated, modified gff from this chunk:
gffRenamedPath <- paste(makerOutput, "mod2_gff.gff", sep = "")
```

Stack overflow - add the contig lines to the mod2_gff.gff file, using the terminal:
$ awk 'p!=$1 {print "# Fasta definition line: >" $1;
print "##sequence-region";
p=$1}1' mod_gff.gff > mod2_gff.gff

After this is done, open the gff in gedit, and add the fasta definition lines 1-7 that are in 
the original blast2go gff.

Generate the mRNA and CDS transcripts using the gff and fasta files:
CDS and Coding Sequence files:
```{r}
getAnnoFastaPath <- paste(scriptsPath, "getAnnoFasta.pl", sep = "")

fnaAssRenamedPath <- paste(makerOutput, 
                           "Renamed_FromB2GSupport_Lsue_assembly_EG_2017_mod_removedlLessThan200_contig00022trim.fna",
                           sep = "")

cmd <- paste("perl ", getAnnoFastaPath,
             " ", gffRenamedPath,
             " --seqfile ", fnaAssRenamedPath, 
             #" --protein=on --codingseq=on ", 
             sep = "")

system(cmd)

codingSeqNtFasta <- paste(makerOutput, "cat Renamed_Blast2GO_gff_without_contig.codingseq", sep = "")
```


This is relevant/necessary:
In-place replacement of names with Maker scripts:
Blast files:
```{r}
# Blast file
blastRenamed <- paste(makerOutput, "Renamed_blast_Lsue_blast2go_export_05Oct2017_1555.txt", sep = "")
modBlastrenamed <- paste(makerOutput, "mod_Renamed_blast_Lsue_blast2go_export_05Oct2017_1555.txt", sep = "")

cmd <- paste("perl ", makerDatMapPath,
             " -m ", modMapFilePath,
             " -i ", blastRenamed,
             " > ", modBlastrenamed,
             sep = "")
system(cmd)
```

Good - Keep:
create blast table in format for maker - make it like the uniprot format:
```{r}
library(data.table)
blastTablePath <- paste(makerOutput, "mod_Renamed_blast_Lsue_blast2go_export_05Oct2017_1555.txt", sep = "")

blastTemp <- fread(blastTablePath, sep = "auto", fill = TRUE)

printrows <- blastTemp[1:6]
printrows$`Blast Top Hit Description (HSP)`
gsub("/.*", "", printrows$`Blast Top Hit Description (HSP)`)
blastTemp$`Blast Top Hit Description (HSP)` <- gsub("/.*", "", blastTemp$`Blast Top Hit Description (HSP)`)
blastTemp$gi <- gsub("^([^\\|]*\\|[^\\|]*)\\|.*$", "\\1", blastTemp$`Blast Top Hit Description (HSP)`)
blastTemp$gi <- gsub("^gi\\|", "", blastTemp$gi)
colnames(blastTemp)
colnames(blastTemp)[1] <- "seqName"

topHitDes <- as.data.table(blastTemp$`Blast Top Hit Description (HSP)`)
topHitDes$GN <- as.data.table(blastTemp$`Sequence Description`)
topHitDes$GN <- gsub("---NA---", "", topHitDes$GN)
topHitDes <- topHitDes[!apply(topHitDes == "", 1, all),]

topHitDes
topHitDes$gi <- as.character(lapply(strsplit(as.character(topHitDes$V1), split = "\\|"), "[",1))
topHitDes$geneID <- as.character(lapply(strsplit(as.character(topHitDes$V1), split = "\\|"), "[",2))
topHitDes$ref <- as.character(lapply(strsplit(as.character(topHitDes$V1), split = "\\|"), "[",3))
topHitDes$accession <- as.character(lapply(strsplit(as.character(topHitDes$V1), split = "\\|"), "[",4))
topHitDes$ProteinName <- as.character(lapply(strsplit(as.character(topHitDes$V1), split = "\\|"), "[",5))
topHitDes$OS <- as.character(lapply(strsplit(as.character(topHitDes$ProteinName), split = "\\["), "[",2))
topHitDes$ProteinName <- gsub("\\[.*", "", topHitDes$ProteinName)
topHitDes$OS <- gsub("]", "", topHitDes$OS)

# Make a list of all the gene ids (gi)
giList <- topHitDes$geneID
write(giList, file = paste(makerOutput, "blast_giList.txt", sep = ""), append = FALSE)
```


Good - Keep:

Need to split the lists to use with UniProt due to connection time-out when there are too many queries.
Use command line script:
$ split -l 100 /home/CFIA-ACIA/girouxeml/PIRL_working_directory/Lachnellula_species_GenomeAn_IonTorrent_2017/Lachnellula_suecica/Lsue_NCBI_files/Lsue_NCBI/Maker_Output/blast_giList.txt

To run idmapping through uniprot on all files in directory, using perl script:
$ for i in *; do perl ~/prog/scripts_pl/uniprot_retrieveIDmapping3.pl $i > $i.txt; done
```{r}
library(data.table)
# Copy the gi's to a new column called "EntryName". We will replace the gi's with the entry names
# retrieved from uniprot's retrieve/idmapping perl script using keys and value matches.
topHitDes$EntryName <- topHitDes$geneID

# Add all the results files from the uniprot perl script idmapping to a list:
tempList <- list.files(path = paste(makerOutput, "split_blast_giLists/", sep = ""), pattern = "*.txt$", full.names = TRUE)

# Create an empty csv to which we will add all the results from the uniprot idmapping perl script:
# Create the empty dataframe
DF <- NULL
# read all of the results files into the empty dataframe 'DF':
for (f in tempList){
    dat <- read.csv(f, header = T, sep = "\t", na.strings="", colClasses = "character")
    DF <- rbind(DF, dat)
}

# convert the 'DF' dataframe to a data table so we can do key-value searches and replacements.
DF <- as.data.table(DF)
nrow(DF) # 12246
nrow(topHitDes) # 12697
# There were no matching entry names in uniprot for 451 gi's.

setkey(topHitDes, EntryName)
setkey(DF, From)
# Replace the gi's in the topHitDes$EntryName column with the entry names of the gi's retieved
# by uniprot idmapping
topHitDes[DF, From := To]
topHitDes[, EntryName:=NULL]
setnames(topHitDes, old = c("From"), new = c("EntryName"))

# If you take a look at the DF and the topHitDes, you will see that not all the gi's matched to uniprot 
# entry names. This is normal: http://www.uniprot.org/help/ncbi_mappings
length(topHitDes$EntryName) # 12697

entryNameList <- topHitDes$EntryName
length(entryNameList) # 12697 

# There are several entries that do not have GIs. These must be removed from the list or the 
# looping later on to combine the results will fail.
is.na(entryNameList)
entryNameList <- na.exclude(entryNameList)
length(entryNameList) # 12299

write(entryNameList, file = paste(makerOutput, "blast_UniProt_entryNameList.txt", sep = ""), append = FALSE)
```


Keep - works!

Make a directory in "Maker_Output" called "split_blast_uniprot_entryName2protname" and cd into it.
Run the following in the shell:
$ split -l 100 /home/CFIA-ACIA/girouxeml/PIRL_working_directory/Lachnellula_species_GenomeAn_IonTorrent_2017/Lachnellula_suecica/Lsue_NCBI_files/Lsue_NCBI/Maker_Output/blast_UniProt_entryNameList.txt

Then run the perl script "uniprot_retrieveIDmapping3.pl" to retrieve the ACC_ID and gene names, and 
protein names of the entries in the the split files in the current directory:
$ for i in *; do perl ~/prog/scripts_pl/uniprot_retrieveIDmapping3.pl $i > $i.txt; done

Some of the results will not have anything in uniprot, even though we were able to retrieve entry names.
When I looked at an entry name manually on the uniprot idmapping page, I saw that no uniprot entry name
was recieved, but there was a match to uniparc. It may be that this is what is happing to the others as
well.

We need to remove the results .txt files that are empty, before continuing to loop and collect the 
results or the loop will fail. This means that the script to retrieve the info from uniprot will
need to be repeated later on to retrieve entries that are in uniparc, but not in uniprot.

In the directory where the results are, run the following in the shell:
$ find . -name '*.txt' -size 0 -print0 | xargs -0 rm

Now only files with >0 bytes remain, continue:
```{r}
library(data.table)
# Collect all the results into a list of results:
tempList2 <- list.files(path = paste(makerOutput, "split_blast_uniprot_entryName2protname/", sep = ""), 
                        pattern = "*.txt$", full.names = TRUE)

# Create an empty data table, with the columns I want:
f <- 1
tempDat <- data.table(entry=character(), geneNames=character(), proteinName=character(), entryName=character())

for (f in tempList2){
    dat2 <- read.table(f, header = F, skip=1, sep = "\t", na.strings="", colClasses = "character", 
                       quote = "", fill = T)
    #dat2 <- read.table(tempList2[1], header = F, skip=1, sep = "\t", na.strings="", colClasses = "character", fill = T)
    setnames(dat2, 1:4,c("entry", "geneNames", "proteinName", "entryName"))
    dat2 <- dat2[ , 1:4]
    tempDat <- rbind(tempDat, dat2)
}

# Now add the uniprot protein and gene names to the topHitsDes data table:
# 1. first add the entry name to new column called "proteinName"
topHitDes$proteinName <- topHitDes$EntryName
topHitDes$geneName <- topHitDes$EntryName

# 2. Set up key-value for renaming the protein names:
setkey(topHitDes, proteinName)
setkey(tempDat, entryName)
topHitDes[tempDat, entryName := proteinName]
topHitDes[, proteinName:=NULL]
setnames(topHitDes, old = c("entryName"), new = c("proteinName"))
```


Keep - works!

Because there are uniparc entries, i have to get the entry names and then get the gene names
by querying both uniprot and uniparc separately, then combining the two lists.

Once I have the gene names, in one list, I can get the rest of the info for the protein names
using the gene names.

Need to get info for uniparc entries.
Go from gi -> ACC+ID (Maker_Output/split_blast_giLists)
Add the ACC+ID (entryName) to topHitDes$EntryName
Collect entry names into a list (entryNameList)
make a new dir called "Maker_Output/ split_blast_uniPARC_entryToGeneName"
split the entryNameList into this directory
Run perl idmapping on entryNameList to get gene names for uniparc entries
go from gene names to protein names, etc.

```{r}
library(data.table)
# In the directory where the results are, run the following in the shell:
# $ find . -name '*.txt' -size 0 -print0 | xargs -0 rm
rm(tempDat2, dat2)
# get a list of all the results files:
tempList3 <- list.files(path = paste(makerOutput, "split_blast_uniPARC_entryToGeneName/", sep = ""), 
                        pattern = "*.txt$", full.names = TRUE)

# Create an empty data table, with the columns I want:
f <- 1

tempDat2 <- data.table(entry=character(), geneNames=character(), entryName=character())

for (f in tempList3){
    dat2 <- read.table(f, header = F, skip = 1, sep = "\t", na.strings="", colClasses = "character", 
                       quote = "", fill = T)
    #dat2 <- read.table(tempList2[1], header = F, skip=1, sep = "\t", na.strings="", colClasses = "character", fill = T)
    setnames(dat2, 1:3,c("entry", "geneNames", "entryName"))
    dat2 <- dat2[ , 1:3]
    tempDat2 <- rbind(tempDat2, dat2)
}
tempDat2$geneNames <- gsub(";.*", "", tempDat2$geneNames)

# Rbind
tempDat1 <- tempDat[,c(1,2,4)]
tempDat3 <- rbind(tempDat2, tempDat1)

# Add the gene names to the topHitDes table using key-value assignment:
# 1. Set up key-value for renaming the gene names:
setkey(topHitDes, EntryName)
setkey(tempDat3, entryName)
topHitDes[tempDat3, geneName := geneNames]

# Capture a list of the gene names to run through uniprot idmapping to retrieve protein names,
# which will include uniparc entries.
geneNameList <- topHitDes$geneName

# There are several entries that do not have GIs. These must be removed from the list or the 
# looping later on to combine the results will fail.
length(geneNameList) # 12697
is.na(geneNameList)

geneNameList <- na.exclude(geneNameList)
length(geneNameList) # 11634

write(geneNameList, file = paste(makerOutput, "blast_UniProt_GeneNameList.txt", sep = ""), append = FALSE)
```


Keep - works!

Make a directory called "split_blast_geneNameList_toUniProt_protname". Cd into it and split the 
geneNameList in this directory. Then run the perl script.

$ split -l 100 /home/CFIA-ACIA/girouxeml/PIRL_working_directory/Lachnellula_species_GenomeAn_IonTorrent_2017/Lachnellula_suecica/Lsue_NCBI_files/Lsue_NCBI/Maker_Output/blast_UniProt_GeneNameList.txt

$ for i in *; do perl ~/prog/scripts_pl/uniprot_retrieveIDmapping5.pl $i > $i.txt; done

Remove all files with 0 bytes.
$ find . -name '*.txt' -size 0 -print0 | xargs -0 rm
```{r}
library(data.table)
# get a list of all the results files:
tempList4 <- list.files(path = paste(makerOutput, "split_blast_geneNameList_toUniProt_protname/", sep = ""), 
                        pattern = "*.txt$", full.names = TRUE)

# Create an empty data table, with the columns I want:
f <- 1
tempDat4 <- data.table(entry=character(), entryName=character(), proteinName=character(), 
                       geneName=character(), proteinEx=character(), seqVers=character())

for (f in tempList4){
    dat4 <- read.table(f, header = T, sep = "\t", na.strings="", colClasses = "character", 
                       quote = "", fill = F)
    cols <- grep("Entry|Entry.name|Protein.names|your|Protein.existence|Version", names(dat4), value = TRUE)
    datTemp <- subset(dat4, select = c(cols))
    colnames(datTemp)
    colnames(datTemp) <- c("entry", "entryName", "proteinName", "proteinEx", "seqVers", "geneName")
    tempDat4 <- rbind(tempDat4, datTemp)
}
colnames(tempDat4)
length(tempDat4$entry)
unique(tempDat4$entryName)
unique(topHitDes$EntryName)
cleanedtempDat4 <- tempDat4[tempDat4$entryName %in% topHitDes$EntryName,]

# Now add the uniprot protein and gene names to the topHitsDes data table:
# 1. Set up key-value for renaming the protein names:
setkey(topHitDes, EntryName)
setkey(cleanedtempDat4, entryName)
topHitDes[cleanedtempDat4, proteinName := proteinName]
topHitDes[cleanedtempDat4, entry := entry]
topHitDes[cleanedtempDat4, proteinEx := proteinEx]
topHitDes[cleanedtempDat4, seqVers := seqVers]

topHitDes$proteinEx <- gsub("Predicted", "4", topHitDes$proteinEx)
topHitDes$proteinEx <- gsub("Inferred from homology", "3", topHitDes$proteinEx)
topHitDes$proteinEx <- gsub("Evidence at protein level", "1", topHitDes$proteinEx)
topHitDes$proteinEx <- gsub("Evidence at transcript level", "2", topHitDes$proteinEx)
unique(topHitDes$proteinEx)
```

Keep - works!
The following will complete the uniprot headers for those entries that successfully returned 
protein names. 
```{r}
library(data.table)
topHitDes$uniProtHeader <- paste(">sp|",topHitDes$entry, "|", topHitDes$EntryName, " ", topHitDes$proteinName,
                                 " OS=", topHitDes$OS, " GN=", topHitDes$geneName,
                                 " PE=", topHitDes$proteinEx, " SV=", topHitDes$seqVers, sep = "")
# Collect those with protein names into a new set:
completeTopHits <- topHitDes[!is.na(topHitDes$proteinName),]
write(completeTopHits$uniProtHeader, 
      file = paste(makerOutput, "sp_completeUniprotHits_db.fa", sep = ""), 
      append = FALSE)
# Prepare the matching blast output for the uniprot hits:
blastSeqNameGI <- subset(blastTemp, select = c("seqName", "gi"))

# check there are common values:
common <- intersect(blastSeqNameGI$gi, completeTopHits$geneID)

# 1. Set up key-value for assigning the sequence names:
completeTopHits$seqName <- NULL
setkey(completeTopHits, geneID)
setkey(blastSeqNameGI, gi)
completeTopHits[blastSeqNameGI, seqName := seqName]

# Add a column that replicates the blast output format:
completeTopHits$blastOutComplete <- paste("sp|", completeTopHits$entry, "|", completeTopHits$EntryName, sep = "")
blastTab <- subset(completeTopHits, select = c("seqName", "blastOutComplete"))

# Replicate the blast output:
write.table(blastTab, 
      file = paste(makerOutput, "sp_completeBlastHits_db.txt", sep = ""), 
      append = FALSE,
      sep = "\t",
      col.names=FALSE, row.names = FALSE, quote = FALSE)
```

Keep - works!
Test to see if Maker works on our replicated uniprot database and blast output:
```{r}
uniprotDB <- paste(makerOutput, "sp_completeUniprotHits_db.fa", sep = "")
blastUniprot <- paste(makerOutput, "sp_completeBlastHits_db.txt", sep = "")

cmd <- paste("perl ", makerFuncGffPath,
             " ", uniprotDB,
             " ", blastUniprot, 
             " ", gffRenamedPath,
             " > ", gffRenamedFuncPath, sep = "")

system(cmd)
```

We have incomplete headers for those that did not retrieve uniprot entry or entry names, and by default
aslo protein names. We need to fix these as well.
```{r}
library(data.table)
# Collect those with no protein names into a new set:
restTopHits <- topHitDes[is.na(topHitDes$proteinName),]

# Those with entry names, but no protein name (i.e., uniPARC matches):
uniParcHits <- restTopHits[!is.na(restTopHits$EntryName)]

# Capture a list of the gene names to run through uniprot idmapping to retrieve protein names,
# which will include uniparc entries.
entryNameList <- uniParcHits$EntryName

# There are several entries that do not have GIs. These must be removed from the list or the 
# looping later on to combine the results will fail.
length(entryNameList) # 3865
is.na(entryNameList)

entryNameList <- na.exclude(entryNameList)
length(entryNameList) # 3865

write(entryNameList, file = paste(makerOutput, "blast_UniParc_entryNameList.txt", sep = ""), append = FALSE)
```
Keep - works!

Make a directory in "Maker_Output" called "split_blast_UniParc_UID" and cd into it.
Run the following in the shell:
$ split -l 100 /home/CFIA-ACIA/girouxeml/PIRL_working_directory/Lachnellula_species_GenomeAn_IonTorrent_2017/Lachnellula_suecica/Lsue_NCBI_files/Lsue_NCBI/Maker_Output/blast_UniParc_entryNameList.txt

Then run the perl script "uniprot_retrieveIDmapping6.pl" to retrieve the ACC_ID and gene names, and 
protein names of the entries in the the split files in the current directory:
$ for i in *; do perl ~/prog/scripts_pl/uniprot_retrieveIDmapping6.pl $i > $i.txt; done

To remove empty files:
In the directory where the results are, run the following in the shell:
$ find . -name '*.txt' -size 0 -print0 | xargs -0 rm

Now only files with >0 bytes remain, continue:
```{r}
library(data.table)
# get a list of all the results files:
tempList5 <- list.files(path = paste(makerOutput, "split_blast_UniParc_UID/", sep = ""), 
                        pattern = "*.txt$", full.names = TRUE)

# Create an empty data table, with the columns I want:
f <- 1
tempDat5 <- data.table(entry=character(), entryName=character(), geneName=character(), 
                       proteinName=character(), proteinEx=character(), seqVers=character(),
                       uniparcID=character())

for (f in tempList5){
    dat5 <- read.table(f, header = T, sep = "\t", na.strings="", colClasses = "character", 
                       quote = "", fill = F)
    cols <- grep("Entry|Entry.name|Gene|Protein.names|Protein.existence|Version|your", names(dat5), value = TRUE)
    datTemp2 <- subset(dat5, select = c(cols))
    colnames(datTemp2)
    colnames(datTemp2) <- c("entry", "entryName", "geneName", "proteinName", "proteinEx", "seqVers", "uniparcID")
    tempDat5 <- rbind(tempDat5, datTemp2)
}
colnames(tempDat5)
length(tempDat5$uniparcID) # 3762
unique(tempDat5$uniparcID)
unique(topHitDes$EntryName)

# There are more than one row for some of the uniparc ID's. We must remove all but the first row for each UID.
cleanedtempDat5 <- unique(tempDat5, by =c("uniparcID"))
cleanedtempDat5 <- cleanedtempDat5[cleanedtempDat5$uniparcID %in% topHitDes$EntryName,]

is.data.table(cleanedtempDat5)
is.data.table(uniParcHits)
summary(cleanedtempDat5)
```

Note: After doing the retrieval using the uniParc IDs (UPI), there are still some entries in the uniParcList
that are not returning information. I looked at these by taking a sample and querying them on the UniProt Retrieve/ID mapping
page, and these show up as having UniProtKB entries that are obsolete. This makes up 104 of those searched
from the uniparc list.
```{r}
# Clean up any set keys:
setkey(topHitDes, NULL)
setkey(cleanedtempDat5, NULL)
setkey(tempDat5, NULL)
setkey(uniParcHits, NULL)

# Can't add values to column that already exists, so working with the dt uniParcHits that has the 
# subset of rows for uniparc entries from the topHitsDes dt, remove the proteinName column:
uniParcHits$proteinName <- NULL
uniParcHits$proteinEx <- NULL
uniParcHits$geneName <- NULL
uniParcHits$seqVers <- NULL
uniParcHits$uniProtHeader <- NULL
uniParcHits$entry <- NULL
uniParcHits$entryName2 <- NULL
uniParcHits$uniparcID <- NULL
names(uniParcHits)

# Now add the uniparc protein and gene names to the uniParcHits data table:
# 1. Set up key-value for renaming the protein names:
setkey(uniParcHits, EntryName)
setkey(cleanedtempDat5, uniparcID)

uniParcHits[cleanedtempDat5, uniparcID := uniparcID]
uniParcHits[cleanedtempDat5, EntryName2 := entryName] # entry name given from uniParc UPI ID.
uniParcHits[cleanedtempDat5, geneName := geneName]
uniParcHits[cleanedtempDat5, proteinName := proteinName]
uniParcHits[cleanedtempDat5, entry := entry]
uniParcHits[cleanedtempDat5, proteinEx := proteinEx]
uniParcHits[cleanedtempDat5, seqVers := seqVers]

# Replace protein existence descriptors with corresponding numeric values:
uniParcHits$proteinEx <- gsub("Predicted", "4", uniParcHits$proteinEx)
uniParcHits$proteinEx <- gsub("Inferred from homology", "3", uniParcHits$proteinEx)
uniParcHits$proteinEx <- gsub("Evidence at protein level", "1", uniParcHits$proteinEx)
uniParcHits$proteinEx <- gsub("Evidence at transcript level", "2", uniParcHits$proteinEx)
unique(uniParcHits$proteinEx)
```

The following will complete the uniparc headers in uniprot format for those entries that successfully returned 
protein names. 
```{r}
library(data.table)
uniParcHits$uniProtHeader <- paste(">sp|",uniParcHits$entry, "|", uniParcHits$EntryName2, " ", uniParcHits$proteinName,
                                 " OS=", uniParcHits$OS, " GN=", uniParcHits$geneName,
                                 " PE=", uniParcHits$proteinEx, " SV=", uniParcHits$seqVers, sep = "")

# Collect those with protein names into a new set:
completeTopHitsUniParc <- uniParcHits[!is.na(uniParcHits$proteinName),]
write(completeTopHitsUniParc$uniProtHeader, 
      file = paste(makerOutput, "sp_completeUniParcHits_db.fa", sep = ""), 
      append = FALSE)


# Prepare the matching blast output for the uniprot hits:
blastSeqNameGI <- subset(blastTemp, select = c("seqName", "gi"))

# check there are common values:
common2 <- intersect(blastSeqNameGI$gi, completeTopHitsUniParc$geneID)

# 1. Set up key-value for assigning the sequence names:
#completeTopHitsUniParc$seqName <- NULL
setkey(completeTopHitsUniParc, geneID)
setkey(blastSeqNameGI, gi)
completeTopHitsUniParc[blastSeqNameGI, seqName := seqName]

# Add a column that replicates the blast output format:
completeTopHitsUniParc$blastOutComplete <- paste("sp|", completeTopHitsUniParc$entry, "|", completeTopHitsUniParc$EntryName2, sep = "")
blastTab2 <- subset(completeTopHitsUniParc, select = c("seqName", "blastOutComplete"))

# Replicate the blast output:
write.table(blastTab2, 
      file = paste(makerOutput, "sp_completeBlastHitsUniParc_db.txt", sep = ""), 
      append = FALSE,
      sep = "\t",
      col.names=FALSE, row.names = FALSE, quote = FALSE)
```

Make a combined blast database text file, and uniprot header database fasta file:
```{r}
blastDb_union <- rbind(blastTab, blastTab2)
blastFastadb_union <- union(completeTopHits$uniProtHeader, completeTopHitsUniParc$uniProtHeader)

# Collect those with protein names into a new set:
write(blastFastadb_union, 
      file = paste(makerOutput, "sp_Union_complete_db.fa", sep = ""), 
      append = FALSE)

# Replicate the blast output:
write.table(blastDb_union, 
      file = paste(makerOutput, "sp_Union_completeBlast_db.txt", sep = ""), 
      append = FALSE,
      sep = "\t",
      col.names=FALSE, row.names = FALSE, quote = FALSE)
```

Keep - works!
Test to see if Maker works on our replicated uniparc/uniprot database and blast output:
Note, I'm adding to the already modified gff, from the first update (prior to repeating only uniparc missing entries)
```{r}
uniprot_uniparc_union <- paste(makerOutput, "sp_Union_complete_db.fa", sep = "")
blast_union_uniprot_uniparc <- paste(makerOutput, "sp_Union_completeBlast_db.txt", sep = "")

gffRenamedFuncPath_union <- paste(makerOutput, "func_Union_Renamed_Blast2GO_gff_without_contig.gff", sep = "")

cmd <- paste("perl ", makerFuncGffPath,
             " ", uniprot_uniparc_union,
             " ", blast_union_uniprot_uniparc, 
             " ", gffRenamedPath,
             " > ", gffRenamedFuncPath_union, sep = "")

system(cmd)
```

See what remains after running uniprot and uniparc id mapping. I still see some that appear to be mappable
from uniprot to uniprot entries, i.e., A0A0F6N2A4_9EURO
```{r}
# Collect those with no protein names into a new set:
restTopHitsUniPa <- uniParcHits[is.na(uniParcHits$uniparcID),]
restTopHitsLast <- restTopHits[is.na(restTopHits$EntryName),]

# Start with retrieving/id mapping the rest from the uniParcTopHits:
# Capture a list of their gene IDs (gi) - seems to work on web query.
giList2 <- restTopHitsUniPa$geneID
length(giList2) # 104
is.na(giList2) # none are NA - good

write(giList2, file = paste(makerOutput, "blast_UniParcRest_giList.txt", sep = ""), 
      append = FALSE)

# Also write a list to search by the EntryName, maybe get better hit with reformatting the perl
# script for querying
entryNameListUniParcRest <- restTopHitsUniPa$EntryName
write(entryNameListUniParcRest, 
      file = paste(makerOutput, "blast_UniParcRest_EntryNameList.txt", sep = ""), 
      append = FALSE)
```

Make a directory to put the results of the id mapping for the rest of the gi's from the unmapped Uniparc results:
```{r}
dir.create(paste(makerOutput, "blast_UniParcRest_giList", sep = ""), showWarnings = TRUE, recursive = FALSE)
file.copy(paste(makerOutput, "blast_UniParcRest_giList.txt", sep = ""), 
          paste(makerOutput, "blast_UniParcRest_giList/", sep = ""))
file.copy(paste(makerOutput, "blast_UniParcRest_EntryNameList.txt", sep = ""), 
          paste(makerOutput, "blast_UniParcRest_giList/", sep = ""))
```
There is no need to split this file, since it only has 104 entries.
Continue by running the perl script:
cd into the new directory and run:
$ perl  ~/prog/scripts_pl/uniprot_retrieveIDmapping7.pl blast_UniParcRest_giList.txt > blast_UniParcRest_giList.txt.txt

The results came back for 17 of the 104 entries:
```{r}
# get a list of all the results files:
tempList5 <- list.files(path = paste(makerOutput, "split_blast_UniParc_UID/", sep = ""), 
                        pattern = "*.txt$", full.names = TRUE)

# Create an empty data table, with the columns I want:
temp6 <- read.table(paste(makerOutput, "blast_UniParcRest_giList/blast_UniParcRest_giList.txt.txt", sep = ""),
                    header = T, sep = "\t", na.strings="", colClasses = "character", quote = "", fill = F)

cols <- grep("Entry|Entry.name|Gene|Protein.names|Protein.existence|Version|your", names(temp6), value = TRUE)
tempDat6 <- subset(temp6, select = c(cols))

# Note all the entries recieved do not have gene names assigned. Omit these and move on to adding
# interpro go annotations.
```

Continue by adding the putative putative gene functions to the fasta files, the ones generated for transcripts
and the ones generated for proteins:
```{r}
exonMakerpath <- paste(makerOutput, "Renamed_Blast2GO_gff_without_contig.cdsexons", sep = "")
transcrMaker <- paste(makerOutput, "Renamed_Blast2GO_gff_without_contig.codingseq", sep = "")
protFna <- paste(makerOutput, "Renamed_Lsue_blast2go_fasta_05Oct2017_1535.fasta", sep = "")
makerFuncFastaPath <- "/home/CFIA-ACIA/girouxeml/prog/maker/maker/src/bin/maker_functional_fasta"

cmd <- paste("perl ", makerFuncFastaPath,
             " ", uniprot_uniparc_union,
             " ", blast_union_uniprot_uniparc,
             " ", protFna,
             " > ", paste(makerOutput, "Renamed_FromB2GSupport_Lsue_assembly_putativeFunction.fna", sep = ""),
             sep = "")
system(cmd)

# Note - this is not correct - there are names already in this file and they become disordered. 
# Look back at the original protein output and see how they are different and what is being updated.
# Also, the blast2go output had names included, much better than my attempts to manually fix them. See
# if it matters if I leave them as is, or what needs to be changed is they have to be made congruent
# between gff and fna files.

# The process is to rename the transcripts and protein fasta files with the uniprot/uniparc headers I made.
```

Below the interPro GO id's and terms are added to the gff that we've just added the top blast hits to:
```{r}
library(data.table)
annotPath2 <- paste(makerOutput, "Lsue_blast2go_annot_30Oct2017.annot", sep = "")
annotTbl2 <- fread(annotPath2, sep = "\t", header=TRUE, fill=TRUE)
# unique(annotTbl2$`Enzyme Codes`)

tblPath <- paste(makerOutput, "blast2go_export_some_30Oct2017.txt", sep = "")
tbl <- fread(tblPath, sep = "\t", header=TRUE, fill=TRUE)
colnames(tbl)
tbl$`Mapping Taxa ID`[1]
#unique(tbl$`InterPro Signatures`)
tblColNames <- c("seqName", "seqLen", "interProName", "interProSignatures",
                 "interProGOid", "interProGOterm", "interProGOcategory", "interProAccn",
                 "interProType", "mappingGeneName", "mappingTaxaID", "mappingXref", 
                 "mappingXrefDB", "mappingGOid", "mappingGOterm", "mappingGOcategory",
                 "annotGOcount", "annotGOid", "annotGOterm", "annotGOcategory")
colnames(tbl) <- tblColNames

# Collect only the rows for which there are values in the column "interProAccn"
# replace the term noIPR in interProAccn with nothing:
tbl$interProAccn <- gsub(pattern = "noIPR", "", tbl$interProAccn)
tblSubset <- tbl[tbl$interProAccn != "",]
tblSubset$annotGOcount <- NULL

# remove tbl from the global environment now that we have the subset of it that we want, 
# becuase tbl is a huge file, and may contribute to system crashes.
rm(tbl)

# Split up the interPro ethod and database id's correctly:
tblSubset$interProSignatures1 <- sapply(strsplit(as.character(tblSubset$interProSignatures),' '), "[", 1)
tblSubset$interProSignatures2 <- sapply(strsplit(as.character(tblSubset$interProSignatures),' '), "[", 2)
tblSubset$interProSignatures2 <- gsub("[[:punct:]]", "", tblSubset$interProSignatures2)


# Split the table into two, one set that has annotGOid values, the other that doesn't but has interPro GO ids
# Set with annotation GO ids:
tblSubAnnotGoids <- tblSubset[tblSubset$annotGOid !="",]
tblAnnieA <- tblSubAnnotGoids[,c("seqName", 
                  "seqName",
                  "interProGOid",
                  "interProSignatures2", 
                  "interProSignatures1",
                  "interProName", 
                  "interProType",
                  "interProGOterm",
                  "mappingGeneName",
                  "mappingXref",
                  "mappingXrefDB",
                  "interProAccn",
                  "annotGOterm",
                  "annotGOid")]
# rename the two columns that differ between tblAnnieA and tblAnnieB:
setnames(tblAnnieA, "interProGOid", "interPro_annotGODid")
setnames(tblAnnieA, "annotGOid", "annot_interProGOid")

# Set without annotation GO ids, but with interpro GO is:
tblSubIPRGoids <- tblSubset[tblSubset$annotGOid =="",]
tblAnnieB <- tblSubIPRGoids[,c("seqName", 
                  "seqName",
                  "annotGOid",
                  "interProSignatures2", 
                  "interProSignatures1",
                  "interProName",
                  "interProType",
                  "interProGOterm", 
                  "mappingGeneName",
                  "mappingXref",
                  "mappingXrefDB",
                  "interProAccn",
                  "annotGOterm",
                  "interProGOid")]

# rename the two columns that differ between tblAnnieA and tblAnnieB:
setnames(tblAnnieB, "annotGOid", "interPro_annotGODid")
setnames(tblAnnieB, "interProGOid", "annot_interProGOid")


# rbind the tables, sort by seqName 
tblAnnie <- rbind(tblAnnieA, tblAnnieB)

# Sort the takble so it's ordered by seqName:
setkey(tblAnnie, "seqName")

# remove tblSubIPRGoids, tblSubset, tblAnnieA and tblAnnieB from the global environment
# as these are huge files and may contribute to system crashes:
rm(tblSubAnnotGoids, tblSubset, tblAnnieA, tblAnnieB)

# write the table as tab delimited text file:
write.table(tblAnnie, 
      file = paste(makerOutput, "annieIPR_table.txt", sep = ""),
      sep = "\t",
      quote = FALSE,
      col.names = FALSE,
      row.names = FALSE,
      append = FALSE)

pathAnnieIPRtbl <- paste(makerOutput, "annieIPR_table.txt", sep = "")
# remove tblAnnie from the global environment. If we want it, we can read it back in since
# we've written it to file:
rm(tblAnnie)
```

Run annie.py in order to get the 3-column table for dbxref, IPR and GO id:
$ python3 ~/prog/annie/genomeannotation-annie/annie.py -ipr annieIPR_table.txt

The keys are off - see issue:
https://github.com/genomeannotation/annie/issues/6 

Need to fix this, since there is no fix or new release:
```{r}
annieTbltoFix <- fread(file = paste(makerOutput, "annie_output.tsv", sep = ""), sep = "\t",
                       header = FALSE)
                       
annieTbltoFix$V2[grepl("GO", annieTbltoFix$V3)] <- "GO"
annieTbltoFix$V2[grepl("InterPro", annieTbltoFix$V3)] <- "InterPro"
annieTbltoFix$V3 <- gsub("InterPro:", "", annieTbltoFix$V3 )

# write the table as tab delimited text file:
write.table(annieTbltoFix, 
      file = paste(makerOutput, "annie_fixed_output.txt", sep = ""),
      sep = "\t",
      quote = FALSE,
      col.names = FALSE,
      row.names = FALSE,
      append = FALSE)

# record the path of the new Annie table:
anniePathFinal <- paste(makerOutput, "annie_fixed_output.txt", sep = "")

# remove the annieTbltoFix since we've written it to file and don't need to keep this
# in the global environment:
rm(annieTbltoFix)
```

Need to update the contig names to the new ones created with Maker, in our Annie Table:
In-place replacement of names with Maker scripts:
```{r}
# I'm choosing to add the GO id's and interPro terms using Maker instead of GAG using Annie's 3-column
# table because I noticed that using GAG creates a gff3/gtf2 hybrid that has quotes, and terminal semilcolons
# and is not consistent with the Maker structure I've been using. It doesn't necessarily mean that is wrong,
# but I think it may create more difficulties.
modAnnieIPRfullRenamed <- paste(makerOutput, "renamed_annieIPR_table.txt", sep = "")
cmd <- paste("perl ", makerDatMapPath,
             " -m ", modMapFilePath,
             " -i ", pathAnnieIPRtbl,
             " > ", modAnnieIPRfullRenamed,
             sep = "")
system(cmd)


# below is the file we need to rename if we are to use the 3-column Annie table
# to add interPro annotations. Note, I used the above with Maker instead.:

# First lets set a path and name for the updated output file we're making of our Annie table:
# modAnnieRenamed <- paste(makerOutput, "renamed_annie_fixed_output.txt", sep = "")
# cmd <- paste("perl ", makerDatMapPath,
#              " -m ", modMapFilePath,
#              " -i ", anniePathFinal,
#              " > ", modAnnieRenamed,
#              sep = "")
# system(cmd)
```


Finally we will add protein domain information to the final annotations using a report from InterProScan. This is down using the following scripts:
ipr_update_gff - adds searchable tags to the gene and mRNA features in the GFF3 files
iprscan2gff3 - adds physical viewable features for daomains that can be displayed in JBrowse, Gbrowse, and Web Apollo.
ipr_update_gff hsap_contig.renamed.putative_function.gff output.renamed.iprscan > hsap_contig.renamed.putative_function.domain_added.gff
iprscan2gff3 output.renamed.iprscan hsap_contig.renamed.gff > visible_iprscan_domains.gff 

```{r}
iprUpdateMaker <- "/home/CFIA-ACIA/girouxeml/prog/maker/maker/src/bin/ipr_update_gff"
gffRenamedFuncDomain <- paste(makerOutput, "func_Union_Domain_Renamed_Blast2GO_gff_without_contig.gff", sep = "")

cmd <- paste("perl ", iprUpdateMaker,
             " ", gffRenamedFuncPath_union,
             " ", modAnnieIPRfullRenamed,
             " > ", gffRenamedFuncDomain,
             sep = "")
system(cmd)



```


##########################################################################################
Extra unwanted:


InterProScan GO annotations:
Rename the contig names in the interPro output for
In-place replacement of names with Maker scripts:
```{r}
iprRenamed <- paste(makerOutput, "Renamed_IPR_Lsue_blast2go_export_05Oct2017_1555.txt", sep = "")
modIPRrenamed <- paste(makerOutput, "mod_Renamed_IPR_Lsue_blast2go_export_05Oct2017_1555.txt", sep = "")

# To see what the files look like:
interPro <- fread(iprPath)
modMap <- fread(modMapFilePath)

cmd <- paste("perl ", makerDatMapPath,
             " -m ", modMapFilePath,
             " -i ", iprRenamed,
             " > ", modIPRrenamed,
             sep = "")
system(cmd)

# Take a look now:
modInterPro <- fread(modIPRrenamed)

# Remove the files to conserve space:
rm(interPro, modMap)
```

Add annotations from IPR and blast to gff file using maker:
This isn't working for some reason, the data from the blast output from blast2go is not formatted 
to work with this properly. It needs to be in UniProt or UniRef format. 

Skip this chunk and see if I can convert the blast data in the xml file to a 
table format, and see if this would work better.
```{r}
gffRenamedFuncPath <- paste(makerOutput, "func_Renamed_Blast2GO_gff_without_contig.gff", sep = "")
uniProtDBPath <- "/home/CFIA-ACIA/girouxeml/Databases/uniprot/uniprot_sprot.fasta"
gffIPRupdate <- paste(makerOutput, "IPR_Renamed_Blast2GO_gff_without_contig.gff", sep = "")

cmd <- paste("perl ", makerIPRupdatePath,
             " ", gffRenamedPath,
             " ", iprRenamed, 
             " > ", gffIPRupdate, sep = "")

system(cmd)

```
